---
title: "Detecting adaptation"
author: "Alice Feurtey"
date: "5/5/2020"
output:
  html_document:
    toc: yes
    toc_float: yes
    css: ~/Documents/epur_css.css
    number_sections: no
    code_folding: hide
  pdf_document:
    toc: yes
logo: ~/Documents/Postdoc_Bruce/Communication/Logo_WWZt.png
---


<br><br>

> Here, I analyse and document my progress with the analysis of a world-wide whole genome sampling of *Zymoseptoria tritici*. Some of the analysis are just exploratory while some other are lined up in a clear path to answering questions related to the history of *Z.tritici* and to better understand its adaptation to various climates.

```{r setup, warning=FALSE, message = FALSE}

library(knitr)
library(reticulate)

#Spatial analyses packages
library(raster)
library(sp)
library(rgdal)
library(maps)
library(geosphere)

#Data wrangling and data viz
library(tidyverse)
library(purrr)
library(RColorBrewer)
library(plotly)
library(cowplot)
library(GGally)
library(corrplot)
library(pheatmap)
library(ggstance)
library('pophelper')
library(ggbiplot)
library(igraph)
library(ggraph)
library(ggtext)
library("UpSetR")
library(scatterpie)


#Pop structure and pop genomic
library(GenomicFeatures)
library(gridExtra)

library(ggtree)
library(tidytree)
library(multcomp)
library(lsmeans)



#GEA
library(lfmm)
library(GWASTools)
library(topGO)


#Statistics
library(car)
library(corrr)
library(lsmeans)
library(multcomp)



#Variables
world <- map_data("world")
project_dir="~/Documents/Postdoc_Bruce/Projects/WW_project/"
lists_dir = "~/Documents/Postdoc_Bruce/Projects/WW_project/WW_PopGen/Keep_lists_samples/"

#Data directories
data_dir=paste0(project_dir, "0_Data/")
metadata_dir=paste0(project_dir, "Metadata/")
fig_dir = "~/Documents/Postdoc_Bruce/Manuscripts/Feurtey_WW_Zt/Draft_figures/"

#Analysis directories
#-___________________
VAR_dir = paste0(project_dir, "1_Variant_calling/")
  depth_per_window_dir = paste0(VAR_dir, "1_Depth_per_window/")
  vcf_dir = paste0(VAR_dir, "4_Joint_calling/")
  mito_SV = paste0(VAR_dir, "6_Mito_SV/")
PopStr_dir = paste0(project_dir, "2_Population_structure/")
  nuc_PS_dir=paste0(PopStr_dir, "0_Nuclear_genome/")
  mito_PS_dir = paste0(PopStr_dir, "1_Mitochondrial_genome/")
Sumstats_dir = paste0(project_dir, "3_Sumstats_demography/")
TE_RIP_dir=paste0(project_dir, "4_TE_RIP/")
   RIP_DIR=paste0(TE_RIP_dir, "0_RIP_estimation/")
   DIM2_DIR=paste0(TE_RIP_dir, "1_Blast_from_denovo_assemblies/")
GEA_dir=paste0(project_dir, "5_GEA/")
fung_dir=paste0(project_dir, "6_Fungicide_resistance/")
virulence_dir = paste0(project_dir, "7_Virulence/")
sel_dir = paste0(project_dir, "8_Selection/")
  gene_list_dir = paste0(sel_dir, "0_Lists_unique_copy/")

#Files
vcf_name="Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.biall_SNP.max-m-1.maf-0.05.thin-1000bp"
vcf_name_nomaf="Ztritici_global_March2021.filtered-clean.AB_filtered.SNP.max-m-0.8.thin-1000bp"
vcf_name_mito = "Ztritici_global_March2021.genotyped.mt.filtered.clean.AB_filtered.variants.good_samples.max-m-80"
Zt_list = paste0(lists_dir, "Ztritici_global_March2021.genotyped.good_samples.args")
gff_file = paste0(data_dir, "Zymoseptoria_tritici.MG2.Grandaubert2015.no_CDS.gff3")
effectors_annotation_file = "~/Documents/Postdoc_Eva/Manuscripts/Accepted/Alice_Cecile_Comparative_genomics/Data_for_publication/Annotations_2018_genomes_for_publication.tab"
#eggnog_annotation = paste0(data_dir, "Zymoseptoria_tritici.MG2.Grandaubert2015.eggnog")
ref_fasta_file = paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa")
metadata_name = "Main_table_from_SQL_Feb_2020"
gene_annotation = read_tsv(paste0(data_dir, "Badet_GLOBAL_PANGENOME_TABLE.txt"))
complete_mito = read_tsv(paste0(data_dir, "Complete_mitochondria_from_blast.txt"), col_names = c("ID_file", "Contig"))

Sys.setenv(PROJECTDIR=project_dir)
Sys.setenv(VARDIR=VAR_dir)
Sys.setenv(VCFDIR=vcf_dir)
Sys.setenv(POPSTR=PopStr_dir)
Sys.setenv(MITOPOPSTR=mito_PS_dir)

Sys.setenv(SUMST=Sumstats_dir)
Sys.setenv(GEADIR=GEA_dir)

Sys.setenv(ZTLIST=Zt_list)
Sys.setenv(GFFFILE = gff_file)
Sys.setenv(REFFILE = ref_fasta_file)
Sys.setenv(VCFNAME=vcf_name)
Sys.setenv(VCFNAME_NOMAF=vcf_name_nomaf)
Sys.setenv(VCFNAME_MITO=vcf_name_mito)

#knitr::opts_chunk$set(echo = F)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
knitr::opts_chunk$set(results = T)


# Metadata and sample lists
##Filtered_samples
filtered_samples = bind_rows(
  read_tsv(paste0(metadata_dir, "Sample_removed_based_on_IBS.args"), col_names = "ID_file") %>%
  mutate(Filter = "IBS"),
read_tsv(paste0(metadata_dir, "Sample_with_too_much_NA.args"), col_names = "ID_file") %>%
  mutate(Filter = "High_NA"),
read_tsv(paste0(metadata_dir, "Samples_to_filter_out.args"), col_names = "ID_file") %>%
  mutate(Filter = "Mutants_etc"))

##Samples in vcf
genotyped_samples = read_tsv(Zt_list, col_names = "ID_file")

## Metadata of genotyped samples 
temp = read_tsv(paste0(metadata_dir, metadata_name, "_Description.tab"), col_names = F) %>% pull()

Zt_meta = read_tsv(paste0(metadata_dir, metadata_name, "_with_collection.tab"), 
                 col_names = temp,
                 na = "\\N", guess_max = 2000) %>%
  unite(Coordinates, Latitude, Longitude, sep = ";", remove = F) %>%
  inner_join(., genotyped_samples)  %>%
  mutate(Country = ifelse(Country == "USA", paste(Country, Region, sep = "_"), Country)) %>%
  mutate(Country = ifelse(Country == "Australia", paste(Country, Region, sep = "_"), Country)) %>%
  mutate(Country = ifelse(Country == "NZ", "New Zealand", Country)) %>%
  mutate(Country = ifelse(Country == "CH", "Switzerland", Country)) %>%
  mutate(Latitude2 = round(Latitude, 2), Longitude2 = round(Longitude, 2))

#genotyped_samples %>%
#  filter(!(ID_file %in% filtered_samples$ID_file)) %>%
#    write_tsv(Zt_list, col_names = F)
eggnog_functions = read_tsv(paste0(data_dir, "eggnog_mapper/query_seqs.fa.emapper.annotations"), col_names = read_tsv(paste0(data_dir, "eggnog_mapper/column_names.tab"), col_names =  c("Ignore", "Names")) %>% pull(Names), comment = "#")



#Define colors
## For continents
#myColors <- c("#04078B", "#a10208", "#FFBA08", "#CC0044", "#5C9D06", "#129EBA","#305D1B")
myColors <- c("#DA4167", "grey", "#ffba0a", "#A20106", "#3F6E0C", "#129eba", "#8fa253" )
names(myColors) = levels(factor(Zt_meta$Continent))
Color_Continent = ggplot2::scale_colour_manual(name = "Continent", values = myColors)
Fill_Continent = ggplot2::scale_fill_manual(name = "Continent", values = myColors)

## For clustering
K_colors = c("#f9c74f", "#f9844a", "#90be6d", "#f5cac3", 
"#83c5be", "#f28482", "#577590", "#e5e5e5", "#a09abc",  "#52796f",
"#219ebc", "#003049", "#82c0cc", "#283618", "white")

## For correlations
mycolorsCorrel<- colorRampPalette(c("#0f8b8d", "white", "#a8201a"))(20)

#Random gradients
greens=c("#1B512D", "#669046", "#8CB053", "#B1CF5F", "#514F59")
blues=c("#08386E", "#1C89C9", "#28A7C0", "#B0DFE8", "grey")

#Bioclim data (from Worldclim)
Bioclim_var = c("Annual Mean Temperature", "Mean Diurnal Range ",
                "Isothermality (BIO2/BIO7) (×100)", "Temperature Seasonality (standard deviation ×100)",
                "Max Temperature of Warmest Month", "Min Temperature of Coldest Month",
                "Temperature Annual Range (BIO5-BIO6)",
                "Mean Temperature of Wettest Quarter","Mean Temperature of Driest Quarter",
                "Mean Temperature of Warmest Quarter", "Mean Temperature of Coldest Quarter",
                "Annual Precipitation", "Precipitation of Wettest Month",
                "Precipitation of Driest Month", "Precipitation Seasonality (Coefficient of Variation)",
                "Precipitation of Wettest Quarter", "Precipitation of Driest Quarter",
                "Precipitation of Warmest Quarter","Precipitation of Coldest Quarter")

pheno_cat = tibble(Phenotype = c(
                     "Max Temperature of Warmest Month", "Mean Temperature of Warmest Quarter",
                     "Min Temperature of Coldest Month", "Mean Temperature of Coldest Quarter",
                     "Precipitation of Driest Month", "Precipitation of Driest Quarter",
                     "Precipitation of Wettest Month", "Precipitation of Wettest Quarter"),
       Category = c("Warmth", "Warmth", "Cold", "Cold",
                    "Drought", "Drought", "Rain", "Rain"))
pheno_table = tibble(BIO_ID = paste0("BIO", c(1:19)),
                     Phenotype = Bioclim_var,
                     Category = c("Temperature", "Temperature range", "Temperature range", "Temperature range",
                                  "Temperature", "Temperature", "Temperature range", "Temperature/rain",
                                  "Temperature/rain", "Temperature", "Temperature", "Precipitation",
                                  "Precipitation", "Precipitation", "Precipitation range", "Precipitation",
                                  "Precipitation", "Temperature/rain", "Temperature/rain"),
                     Sub_category = c("Annual", "Annual", NA, "Annual",
                                      "High", "Low", "Annual", NA,
                                      NA, "High", "Low", "Annual",
                                      "High", "Low", "Annual", "High",
                                      "Low", NA, NA))


#Bioclim_var[c(2, 3, 7)] = "Ranges bioclimatic variables"
#Bioclim_var[c(4, 15)] = "Seasonality bioclimatic variables"
#Bioclim_var[c(1, 12) = "Annual rain and temperature"
#Bioclim_var[c(8, 9, 17, 18)] = "Mix rain and temperature" 

```





#GEA: all data
***
## Geographic/environmental data
I will use the sampling site of each isolate (as precisely as I can manage) to approximate environmental parameters such as temperature, or precipitation. One possibility to find such data is this website, https://www.worldclim.org/data/worldclim21.html (published in Fick and Hijmans, 2017), which gives access to climate data for 1970-2018. These can be transformed into bioclimatic variables using the biovars function from the R package dismo (https://rdrr.io/cran/dismo/man/biovars.html).

```{r Get Bioclim data}

temp = Zt_meta %>%
  #filter(!(ID_file %in% filtered_samples$ID_file)) %>%
  filter(!is.na(Longitude)) %>%
  mutate(X = as.numeric(unlist(Longitude)),
                            Y = as.numeric(unlist(Latitude))) %>%
  dplyr::select(X, Y) %>%
  distinct()

sp = SpatialPoints(temp[, c("X", "Y")])
summary(sp)



bio_list = list()
for (i in c(1:length(Bioclim_var))) {
  file_name=paste0(data_dir,"Climatic_data/wc2.1_10m_bio/wc2.1_10m_bio_", i, ".tif")
  rast_temp = raster(file_name)    
  bio_list[[Bioclim_var[i]]] = raster::extract(rast_temp, sp)
}

climate_per_coordinates = cbind(temp, do.call(cbind, bio_list))

```

```{r Plot distrib bioclim}

dat = Zt_meta %>%
  #filter(!(ID_file %in% filtered_samples$ID_file)) %>% 
  dplyr::select(ID_file, Latitude, Longitude, Continent) %>%
  full_join(., climate_per_coordinates,
                by= c("Longitude" = "X", "Latitude" = "Y")) %>%
  dplyr::select(-ID_file) %>%
  gather(key = "Bioclim_var", value = "Estimate", -c(Longitude, Latitude, Continent))

#Define stable colors
#myColors = c("#ec9a29", "#0f8b8d", "#143642")
#names(myColors) = levels(factor(dat$Continent))
#colScale = scale_colour_manual(name = "Continent", values = myColors)
#colScaleFill = scale_fill_manual(name = "Continent", values = myColors)

ggplot(dat, aes(Estimate, fill =Continent, color =Continent)) +
  geom_histogram(position = "stack") +
  facet_wrap(.~Bioclim_var, scales = "free") +
  theme_classic() + Color_Continent + Fill_Continent


```

Naturally, many of the variables investigated above are highly correlated. It is intuitive for example that the minimum temperature of the coldest month would be correlated to the average temperature of the coldest quarter! Here, I visualize these correlations.

```{r correl climate var}

#Correlogram

Ccor = cor(climate_per_coordinates[, c(3:ncol(climate_per_coordinates))])
corrplot(Ccor, type="lower", order="hclust",
         col = mycolorsCorrel,
         tl.col="black", tl.srt=45, tl.cex = 0.7)

#Simple heatmap
heatmap(x = Ccor, col = mycolorsCorrel, symm = TRUE)

Ccor[abs(Ccor) <= 0.75] <-0
heatmap(x = Ccor, col = mycolorsCorrel, symm = TRUE)
```



```{r correl climate net}


tidy_cors <- climate_per_coordinates %>%
  dplyr::select(everything(), -contains("uarter"), -Y, -X) %>%
  correlate() %>%
  stretch() 

graph_cors <- tidy_cors %>%
  #filter(abs(r) > .2) %>%
  graph_from_data_frame(directed = FALSE)

ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = .5, color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(-1, 1), colors = c("#0f8b8d", "white", "#a8201a")) +
  #geom_node_point(color = "black", size = 3) +
  geom_node_text(aes(label = name), size = 2) +
  theme_graph() +
  labs(title = "Correlations between bioclimatic variables")

#
tidy_cors <- climate_per_coordinates %>%
  dplyr::select( -Y, -X) %>%
  correlate() %>%
  stretch() 

graph_cors <- tidy_cors %>%
  #filter(abs(r) > .75) %>%
  mutate(r = ifelse(abs(r) <= 0.75, 0, r)) %>%
  graph_from_data_frame(directed = FALSE)

ggraph(graph_cors) +
  geom_edge_link(aes(edge_alpha = .5, color = r)) +
  guides(edge_alpha = "none", edge_width = "none") +
  scale_edge_colour_gradientn(limits = c(-1, 1), colors = c("#0f8b8d", "white", "#a8201a")) +
  #geom_node_point(color = "black", size = 3) +
  geom_node_text(aes(label = name), size = 2) +
  theme_graph() +
  labs(title = "Correlations between bioclimatic variables")

```


```{r standard Bioclim var}

#Create standardized values for the environment variables

climate_per_coord_standard = climate_per_coordinates %>%
  pivot_longer(cols = -c(X, Y), names_to = "Variable_climate", values_to = "Value") %>%
  group_by(Variable_climate) %>%
  mutate(Standard_value = scale(Value)) %>%
  dplyr::select(-Value) %>%
  pivot_wider(names_from = Variable_climate, values_from = Standard_value)
```

```{r GEMMA prep climate file, eval = F}

#The fam file should be same for all core chromosomes, hopefully. So I only have to create one to get the file format including the phenotypes. This can then be used for all chromosomes.

temp2 = left_join(read_tsv(Zt_list, col_names = "ID_file") %>% mutate(ID2 = ID_file), 
                            Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude)) %>%
  bind_cols(., tibble(X1 = rep(0, nrow(Zt_meta)), X2 = rep(0, nrow(Zt_meta)), X3 = rep(0, nrow(Zt_meta)))) %>%
  left_join(., climate_per_coord_standard,
                   by = c("Latitude" = "Y", "Longitude" = "X")) 

dplyr::select(temp2, -Latitude, -Longitude) %>%
 write_delim(., paste0(GEA_dir, "Ztmeta.fam"), 
               delim = " ", col_names = F)

left_join(read_tsv(Zt_list, col_names = "ID_file") %>% mutate(ID2 = ID_file), 
                            Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude)) %>%
  bind_cols(., tibble(X1 = rep(0, nrow(Zt_meta)), X2 = rep(0, nrow(Zt_meta)), X3 = rep(0, nrow(Zt_meta)))) %>%
  left_join(., climate_per_coordinates,
                   by = c("Latitude" = "Y", "Longitude" = "X")) %>%
  dplyr::select(-Latitude, -Longitude) %>%
 write_delim(., paste0(GEA_dir, "Ztmeta_non_standard.fam"), 
               delim = " ", col_names = F)
#Transfer on the cluster
#rsync -avP ~/Documents/Postdoc_Bruce/Projects/WW_project/5_GEA/Ztmeta.fam alice@130.125.25.244:/data2/alice/WW_project/5_GEA/Phenotypes_with_climatic_variables.fam

#To run gemma on the cluster: conda activate env0
#Commands are then in the format gemma -h 
```


## Identify variants significantly associated to bioclimatic variables

First, I read the results that I got from running GEMMA with a LOCO strategy (i.e., leave-one-chromosome-out). 
```{r GEMMA results}

results_GEMMA = list()
for (i in c(1:length(Bioclim_var))) {
  temp = list.files(path = paste0(GEA_dir, "0_Basic/"), pattern = paste0("*phenotype_", i,".subset_ALL.assoc.txt")) %>% 
    map_df(~read_tsv(file = paste0(GEA_dir, "0_Basic/", .), col_types = c("dcddccdddddd"))) %>%
    unite(col = SNP, chr, ps, sep = "_", remove = F) %>%
    dplyr::select(SNP, CHR = chr, BP = ps, P = p_wald, zscore = logl_H1) %>%
    mutate(Phenotype = Bioclim_var[i], Nb = i, Subset = "ALL")
  
  results_GEMMA[[Bioclim_var[i]]] = temp
}

results_GEMMA = bind_rows(results_GEMMA)


```

Once I have the data loaded, it's time for some sanity checks. Here, I will look at the genomic inflation factor as well as some qqplots. 

```{r checks GWAS}
#Genomic Inflation Factor
temp = results_GEMMA %>% 
        group_by(Phenotype) %>%
        summarize(Genomic_Inflation_Factor = median(qchisq(1-P,1))/qchisq(0.5,1))
BIO_sum_table =  full_join(pheno_table, temp) %>% arrange(Genomic_Inflation_Factor)       
kable(BIO_sum_table)

#QQ-plots
par(mfrow=c(3,3)) 
for (i in c(1:length(Bioclim_var))){
temp = results_GEMMA %>% filter(Phenotype == Bioclim_var[i]) %>% dplyr::sample_frac(size = .10) %>% dplyr::select(P) %>% pull()
GWASTools::qqPlot(temp, thinThreshold = 2)
}



```
Both sets of checks indicate that there are some variants with higer p-values than expected. The values are a bit high, but this could be expected in the case of a polygenic trait, and since I have no reason to expect that local adaptation to climatic variable should be monogenic, the values are acceptable.

To determine which variants are significant, we use the Bonferroni correction method, i.e. we divide the traditional threshold of significance of 0.05 by the number of variants that we tested. Based on this corrected threshold, we can then identify variants significantly associated with each environmental condition that we tested.
```{r GWAS signif}

Bonferroni_thresholds = results_GEMMA %>% dplyr::count(Phenotype, Nb) %>%
  mutate(Bonferroni_threshold = 0.05/n) #%>% summarize(average = mean(Bonferroni_threshold)) %>% pull()
results_GEMMA = full_join(results_GEMMA, Bonferroni_thresholds)

significant = results_GEMMA %>%
  filter(P <= Bonferroni_threshold)  %>% group_by(SNP, CHR, BP) %>%
  mutate(Count = n()) %>% mutate(SNP_type = "significant")%>% 
  ungroup()

BIO_sum_table = full_join(BIO_sum_table, results_GEMMA) %>% 
  filter(P <= Bonferroni_threshold) %>% 
  dplyr::count(Phenotype) %>% 
  arrange(n) %>%
  dplyr::select(Phenotype, Nb_signif_variants = n)

significant %>%
  dplyr::select(CHR, BP) %>% 
  distinct() %>%
  write_delim(paste0(GEA_dir, "Significant_GEMMA.tsv"), delim = "\t", col_names = F)
```

Identifying the variants which are significant is very useful already, but it is hard to work with so many "independent" positions. It also does not make sense conceptually, as nearby variants are probably picking up the exact same signal through linkage. So, here I gather variants together into "significant loci" if there are no bigger gaps than a certain threshold.
```{r signif loci}

#Function to gather positions into loci
from_positions_to_loci <- function(tibble_pos, distance_threshold) {
  chromosomes = unique(sort(tibble_pos$CHR))
  temp2 = list()
  for (i in chromosomes){
    v = tibble_pos %>%
      filter(CHR == i) %>%
      dplyr::select(CHR, BP) %>% 
      distinct() %>% pull(BP) %>% sort()
    
    # Split the vectors into group based on the length threshold
    temp = split(v, cumsum(c(TRUE, diff(v) >= distance_threshold)))
    
    temp2[[i]] =  temp %>% map_df(enframe, .id = 'Locus_nb') %>% rename(BP = value) %>%
      mutate(CHR = i)
    
  }
  
  bind_rows(temp2) %>%
    dplyr::select(CHR, Locus_nb, BP) %>%
    arrange(CHR, Locus_nb, BP)
}


distance_threshold = 10000L #Distance between two significant SNPs to include them in the same region

#For each phenotype, get significant loci from list of positions
list_loci = list()
for (j in c(1:length(Bioclim_var))) {
  list_loci[[j]] = from_positions_to_loci(filter(significant, Phenotype == Bioclim_var[j]),
                         distance_threshold) %>%
    mutate(Phenotype = Bioclim_var[j])
}

#Adding locus info to the table
significant = full_join(bind_rows(list_loci), significant) %>% 
  group_by(CHR, Locus_nb, Phenotype) %>%
  mutate(Locus_length = 1 + max(BP) - min(BP)) %>%
  ungroup()


BIO_sum_table = full_join(BIO_sum_table, significant %>% dplyr::select(Locus_nb, Phenotype) %>%
  distinct() %>% 
  dplyr::count(Phenotype) %>%
  dplyr::select(Phenotype, Nb_signif_loci = n))
kable(BIO_sum_table)

ggplot(significant, aes(Locus_length)) + 
  geom_density(fill = "#82c0cc", alpha =.4, col = "#82c0cc") +
  theme_bw() +
  labs(title = "Distribution of the length of significant loci", 
       subtitle = paste0("Significant variants were grouped together if they were closer than ", 
                                  distance_threshold, " bp."),
       x = "Length (bp)")
```


## Minor allele frequencies of significant variants
In many different GWAS, it has been found that significant variants are often found at low minor allele frequencies. My threshold to keep a variant in the analysis was initially 0.01, but often it is 0.05. Here, I want to see how many of the significant variants we have found previously as significant fall below/above 0.05. 

```{r signif MAF}
#Read MAF data for the significant alleles
temp = read_tsv(paste0(GEA_dir, "Significant_GEMMA.ann.tab"), na = ".") %>%
  unite(col = SNP, CHROM, POS) %>%
  pivot_longer(-c(REF, ALT, SNP), names_to = "ID_file", values_to = "Allele") %>%
  group_by(SNP) %>%  
  mutate(Allele = ifelse(Allele == 0, "REF", "ALT")) %>%
  filter(Allele != "NA") %>%
  dplyr::count(Allele) %>%
  pivot_wider(names_from = Allele, values_from = n, values_fill = 0) %>%
  mutate(Nb_genotypes = ALT + REF,
         REF_prop = REF/Nb_genotypes, ALT_prop = ALT/Nb_genotypes, 
         MAF = min(REF_prop, ALT_prop)) %>%
  dplyr::select(-ALT, -REF)


significant = left_join(significant, temp)

#Violin plot MAF
ggplot(data = significant, aes(x = Phenotype, y = MAF, col = Phenotype, fill = Phenotype)) +
  geom_violin(alpha = .5) +
  theme_bw() + 
  theme(legend.position = "none") + 
  scale_y_continuous(trans='log2') +
  coord_flip() +
  labs(title = "Distribution of the MAF of significant variants",
       subtitle = "Beware! Logarithmic scale.",
       x = "")


# Basic barplot
data <- significant %>% 
  mutate(group = ifelse(MAF < 0.05, "1to5%", ">5%")) %>%
  filter(!is.na(group))

temp = dplyr::count(data, Phenotype, group)

ggplot(data, aes(x=Phenotype, fill=group)) +
  geom_bar(position = "fill") +
  coord_flip() +
  theme_bw() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) + 
  scale_fill_manual(values = c("#82c0cc", "#EDE7E3")) +
  labs(y = "Proportion of significant variants", 
       fill = "MAF category", x = "") +
  geom_label(data = temp %>% filter(group == ">5%"),
            aes(label = n, y = 1.13), colour = "white")  +
  ylim(c(0, 1.2)) +
  labs(title = str_wrap("Proportion and numbers of significant variants with a MAF > 5%", 40))


```
As expected a large proportion of variants with a significant association to one of the bioclimatic variable have a low MAF. I would like to put this in perspective by comparing with variants from the pool of tested variants, that is biallelic variants without any missing data and 
```{r maf comparison}
prefix = "Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.max-m-100.biall"
biall_maf_ann = full_join(read_tsv(paste0(vcf_dir, prefix, ".frq.count"), 
                                   col_names = c("CHR", "BP", "N_ALL", "Total", "REF_count", "ALT_count"),
                                   skip = 1) %>%
                            rowwise() %>%
                            mutate(CHR = as.numeric(CHR), BP = as.numeric(BP),
                                   MAF = min(as.numeric(REF_count), as.numeric(ALT_count))/as.numeric(Total)) %>%
                            dplyr::select(-REF_count, -Total, -ALT_count, -N_ALL),
                          read_tsv(paste0(vcf_dir,prefix, ".ann.tsv"), 
                                   col_names = c("CHR", "BP", "REF", "ALT", "ANN")) %>% 
                            separate(ANN, sep = ",", 
                                     into = c("ANN1", "ANN2", "ANN3", "ANN4", "ANN5", "ANN6", "ANN7", 
                                              "ANN8", "ANN9", "ANN10", "ANN11",
                                              "ANN12", "ANN13", "ANN14", "ANN15", "ANN16", "ANN17"), 
                                     extra = "warn")) %>%
  filter(CHR != "mt") %>%
  mutate(CHR = as.numeric(CHR), BP = as.numeric(BP)) %>%
  full_join(., significant %>% dplyr::select(CHR, BP, SNP_type))  %>%
  mutate(SNP_type = ifelse(is.na(SNP_type), "All variants", SNP_type))%>%
  filter(MAF >= 0.01) 



ggplot(biall_maf_ann, aes(MAF, fill = SNP_type)) + 
    geom_density(alpha = .4) + 
    facet_wrap(vars(SNP_type), scales = "free") + 
    theme_bw() +
  theme(legend.position = "none") +
  labs(title = "Comparison between MAF of all variants and significant variants only")

```

## Functional effect of the significant variants

Each SNP also has a predicted effect on the protein sequences it affects. snpEff groups the effect into 4 broad categories which are: 

* high, e.g. a nonsense mutation, 
* moderate, e.g. a missense mutation, 
* low, e.g. as a synonymous mutation,
* modifier, mutations at 1kb of a gene (threshold chosen based on Zt parameters, default is 5kb)

Provided we do have the causal variants in our GEMMA analysis, we could expect that variants with a stronger effect on the protein sequence would have higher p-values than synonymous variants for instance. Alternatively, if the climate adaptation response involves mistly changes in regulation, then we would instead expect a lot of modifier mutations to have high p-values. 
```{r snpeff vs logP}


#Translating SnpEff effects in a numerical scale
gene_detected = inner_join(pheno_table, significant) %>%
    mutate(logP = -log10(P)) %>%
    left_join(., signif_ann) %>% 
  mutate(Ranked_effect_strength = Effect_strength)
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "HIGH"] = 1
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODERATE"] = 2
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "LOW"] = 3
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODIFIER"] = 4
gene_detected$Ranked_effect_strength[is.na(gene_detected$Effect_strength)] = 5
gene_detected$Effect_strength[is.na(gene_detected$Effect_strength)] = "NO_EFFECT"

#Translating SnpEff effects in a numerical scale
gene_detected = gene_detected %>% 
  mutate(Ranked_effect_strength = Effect_strength)
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "HIGH"] = 1
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODERATE"] = 2
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "LOW"] = 3
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODIFIER"] = 4
gene_detected$Ranked_effect_strength[is.na(gene_detected$Ranked_effect_strength)] = 5
gene_detected$Effect_strength[is.na(gene_detected$Effect_strength)] = "NO_EFFECT"


#For all variant, 
#selecting the highest effect 
temp = gene_detected %>%
  #filter( MAF >= 0.05 ) %>%
  group_by(SNP, Effect_strength, Ranked_effect_strength) %>% 
  summarize(logP = max(logP)) %>% group_by(SNP) %>% 
  mutate(ranktemp = rank(Ranked_effect_strength, ties.method = "random")) %>%
  filter(ranktemp == 1) 

#Statistical test
#One-way ANOVA with blocks
##Define linear model
model = lm(as.numeric(logP) ~ Effect_strength ,
          data=temp)

##Conduct analysis of variance
Anova(model,type = "II")  

#Post-hoc analysis:  mean separation tests
marginal = lsmeans(model, ~ Effect_strength)
pairs(marginal, adjust="tukey")

CLD_dim = cld(marginal, alpha   = 0.05,
          Letters = letters,  ### Use lower-case letters for .group
          adjust  = "tukey")  ### Tukey-adjusted p-values

CLD_dim
CLD_dim$.group=gsub(" ", "", CLD_dim$.group)
text_y = (max(as.numeric(temp$Composite_median), na.rm = T) + 0.1*max(as.numeric(temp$Composite_median), na.rm = T))

#Plot the log p-value for each of the snpEff cateogory
max_val = 30
dplyr::group_by(temp, SNP, Effect_strength) %>% 
  group_by(Effect_strength) %>%
  mutate(ave_logP = median(logP)) %>%
  filter(logP < max_val) %>%
  ggplot(aes(x = fct_reorder(Effect_strength, ave_logP), y = logP)) + 
     geom_jitter(width = .25, fill = NA, shape = 1, col = "#82C0CC")+ 
     geom_boxplot(col = "#489FB5", outlier.shape = NA, fill = NA)  +
     theme_bw() +
     geom_text(data = CLD_dim, aes(x = Effect_strength,
                            y = max_val,
                            label = .group), color = "black") +
     geom_text(data = as.data.frame(table(temp$Effect_strength)), 
               aes(x = Var1,
                            y = max_val + 2,
                            label = paste0("N=", Freq)), color = "grey30")+
  labs(x = "Effect of the mutation on protein (snpEff)",
       y = "Maximum -log10(p-value) in any GEA",
       title = str_wrap("Comparison of the p-values of variants with different levels of effect on the proteic sequence", 80),
       subtitle = str_wrap(paste0("Dots represent individual variants and the boxplots summarize the distribution.",
                                  "The letters indicate the statiscally distinct groups of distribution, ",
                                  "while the numbers of the top are the count of variants in each category."), 100)) 

```
While taking all significant variants into account, there is no statistical difference between the different groups. Let's see if the absence of difference holds true when considering only the variants with a MAF higher than 5%. 

```{r snpeff maf > 5}


#For all variant, 
#selecting the highest effect 
temp = gene_detected %>%
  filter( MAF >= 0.05 ) %>%
  group_by(SNP, Effect_strength, Ranked_effect_strength) %>% 
  summarize(logP = max(logP)) %>% group_by(SNP) %>% 
  mutate(ranktemp = rank(Ranked_effect_strength, ties.method = "random")) %>%
  filter(ranktemp == 1) 

#Statistical test
#One-way ANOVA with blocks
##Define linear model
model = lm(as.numeric(logP) ~ Effect_strength ,
          data=temp)

##Conduct analysis of variance
Anova(model,type = "II")  

#Post-hoc analysis:  mean separation tests
marginal = lsmeans(model, ~ Effect_strength)
pairs(marginal, adjust="tukey")

CLD_dim = cld(marginal, alpha   = 0.05,
          Letters = letters,  ### Use lower-case letters for .group
          adjust  = "tukey")  ### Tukey-adjusted p-values

CLD_dim
CLD_dim$.group=gsub(" ", "", CLD_dim$.group)
text_y = (max(as.numeric(temp$Composite_median), na.rm = T) + 0.1*max(as.numeric(temp$Composite_median), na.rm = T))

#Plot the log p-value for each of the snpEff cateogory
max_val = 30
dplyr::group_by(temp, SNP, Effect_strength) %>% 
  group_by(Effect_strength) %>%
  mutate(ave_logP = median(logP)) %>%
  filter(logP < max_val) %>%
  ggplot(aes(x = fct_reorder(Effect_strength, ave_logP), y = logP)) + 
     geom_jitter(width = .25, fill = NA, shape = 1, col = "#82C0CC")+ 
     geom_boxplot(col = "#489FB5", outlier.shape = NA, fill = NA)  +
     theme_bw() +
     geom_text(data = CLD_dim, aes(x = Effect_strength,
                            y = max_val,
                            label = .group), color = "black") +
     geom_text(data = as.data.frame(table(temp$Effect_strength)), 
               aes(x = Var1,
                            y = max_val + 2,
                            label = paste0("N=", Freq)), color = "grey30")+
  labs(x = "Effect of the mutation on protein (snpEff)",
       y = "Maximum -log10(p-value) in any GEA",
       title = str_wrap(paste0("Comparison of the p-values of variants with different ",
                               "levels of effect on the proteic sequence and a MAF > 0.05"), 80),
       subtitle = str_wrap(paste0("Dots represent individual variants and the boxplots summarize the distribution.",
                                  "The letters indicate the statiscally distinct groups of distribution, ",
                                  "while the numbers of the top are the count of variants in each category."), 100)) 
```
This time there is a significant difference between the maximum logP per variant. 

I now want to compare the distributions of effects in significant vs all variants. Are there more high effect variants in the significant variants than expected at random? 

```{r effect comparison}

signif_ann = right_join(biall_maf_ann, significant %>% dplyr::select(CHR, BP))  %>%
  pivot_longer(cols = c(-CHR, -BP, -REF, -ALT, -MAF, -SNP_type), names_to = "ANN", values_to = "Annotation") %>%
  filter(!is.na(Annotation)) %>%
  separate(Annotation, into = c("ALT2", "Variant_type", "Effect_strength", "Gene", "Rest"), sep = "\\|", extra = "merge") %>%
  filter(!str_detect(Gene, "-")) %>%
  separate(Gene, into = c("temp", "Chrom", "Gene_nb"), remove = FALSE)  %>%
  dplyr::select(-ANN, -ALT2, -temp) %>%
  mutate(Gene_nb = as.numeric(Gene_nb)) %>%
  left_join(., read_tsv(effectors_annotation_file) %>% 
              filter(Sample == "Zt09") %>% 
              mutate(Gene = str_replace(Protein_ID, "_chr", ""))) %>%
  mutate(Ranked_effect_strength = Effect_strength)
signif_ann$Ranked_effect_strength[signif_ann$Ranked_effect_strength == "HIGH"] = 1
signif_ann$Ranked_effect_strength[signif_ann$Ranked_effect_strength == "MODERATE"] = 2
signif_ann$Ranked_effect_strength[signif_ann$Ranked_effect_strength == "LOW"] = 3
signif_ann$Ranked_effect_strength[signif_ann$Ranked_effect_strength == "MODIFIER"] = 4
signif_ann$Ranked_effect_strength[is.na(signif_ann$Effect_strength)] = 5
signif_ann$Effect_strength[is.na(signif_ann$Effect_strength)] = "NO_EFFECT"


effect_counts_signif = signif_ann %>%
    group_by(CHR, BP) %>%
    mutate(Rank = rank(Ranked_effect_strength, ties.method = "first")) %>%
    arrange(Rank) %>%
    ungroup() %>%
    filter(Rank == 1) %>% 
    mutate(MAF_group = ifelse(MAF < 0.05, "1to5%", ">5%")) %>%
    dplyr::count(Effect_strength) %>%
    mutate(Total = sum(n), Percent = 100*n/Total)

temp1 = biall_maf_ann %>% filter(MAF >= min_maf_signif)
nb = sum(effect_counts_signif$n)
effect_counts = list()
for (i in c(1:100)) {
  temp = dplyr::slice_sample(as.tibble(temp1), n  = nb) %>%
    pivot_longer(cols = c(-CHR, -BP, -REF, -ALT, -MAF, -SNP_type), names_to = "ANN", values_to = "Annotation") %>%
    filter(!is.na(Annotation)) %>%
    separate(Annotation, into = c("ALT2", "Variant_type", "Effect_strength", "Gene", "Rest"), sep = "\\|", extra = "merge") %>%
    filter(!str_detect(Gene, "-")) %>%
    separate(Gene, into = c("temp", "Chrom", "Gene_nb"), remove = FALSE)  %>%
    dplyr::select(-ANN, -ALT2, -temp) %>%
    mutate(Gene_nb = as.numeric(Gene_nb)) %>%
    mutate(Ranked_effect_strength = Effect_strength)
    
  temp$Ranked_effect_strength[temp$Ranked_effect_strength == "HIGH"] = 1
  temp$Ranked_effect_strength[temp$Ranked_effect_strength == "MODERATE"] = 2
  temp$Ranked_effect_strength[temp$Ranked_effect_strength == "LOW"] = 3
  temp$Ranked_effect_strength[temp$Ranked_effect_strength == "MODIFIER"] = 4
  temp$Ranked_effect_strength[is.na(temp$Effect_strength)] = 5
  temp$Effect_strength[is.na(temp$Effect_strength)] = "NO_EFFECT"
  
  effect_counts[[i]] = temp %>%
    group_by(CHR, BP) %>%
    mutate(Rank = rank(Ranked_effect_strength, ties.method = "first")) %>%
    arrange(Rank) %>%
    ungroup() %>%
    filter(Rank == 1) %>%
    mutate(repet = paste0("repet_", i)) %>%
    dplyr::count(Effect_strength) %>%
    mutate(Total = sum(n), Percent = 100*n/Total)
}
rm(temp1)
effect_counts = bind_rows(effect_counts)


p1 = ggplot() +
  geom_jitter(data = effect_counts, aes(x = Effect_strength, y = Percent), 
              shape = 1, alpha = .8, col = "#DFD4CD") +
  geom_point(data = effect_counts_signif, aes(x = Effect_strength, y = Percent), 
             shape = 19, size = 2, col = "#82C0CC") +
  theme_bw() +
  labs(y = "Percentage of variant from each category", x = "")

p2 = ggplot(data = effect_counts) +
  geom_density(aes(x = Percent), col = "#DFD4CD", alpha = .4, fill = "#EDE7E3") +
  geom_vline(data = effect_counts_signif, aes(xintercept = Percent), col = "#82c0cc") +
  theme_bw() +
  facet_wrap(vars(Effect_strength), scales = "free")+
  labs(x = "Percentage of variant from each category", y = "Density")
row = cowplot::plot_grid(p1, p2)

title_gg <- ggplot() + 
  labs(title = str_wrap(paste0("Comparison between effect proportions in significant ",
                               "variants vs randomly selected variants (with MAF > 0.01)"), 80))
cowplot::plot_grid(title_gg, row, nrow = 2, rel_heights = c(0.15, 1))

```

## Functional significance

```{r}
# Table with genes affected by significant variants with high MAF
# with all the functional annotations
temp = gene_detected %>% 
  filter( MAF > 0.05 ) %>%
  filter(Category %in% c("Temperature", "Precipitation"))  %>%
  filter(Sub_category != "Annual") %>%
  dplyr::select("Phenotype","Category","Sub_category", "SNP" , "Variant_type","Effect_strength",
                "Gene", "Secretion", "Transmembrane", "Cell_location", "Cluster(nb, function)", "Effector", "CAZyme", 
                "Summary_RNAseq") %>%
  left_join(eggnog_functions, by = c("Gene" = "query_name"))
```

### Functional enrichment
The genes found in each of the GWAS could have some type of enrichment. 
```{r GO enrichment}

geneNames = read_tsv(gff_file, 
               col_names = c("CHR", "X1", "mRNA", "Start", "Stop", "X2", "X3", "X4", "Annot"))  %>%
  filter(mRNA == "mRNA") %>%
  separate(col = Annot, into = c("ID", "Parent", "Name"), sep = ";") %>%
  mutate(Name = str_remove(Name, "Name=")) %>%
  pull(Name)
geneID2GO = readMappings(paste0(data_dir, "GO_Zt09.map"))
gene_detected = inner_join(pheno_table, significant) %>%
    mutate(logP = -log10(P)) %>%
    left_join(., signif_ann)# %>%
  #  filter(Effect_strength %in% c("HIGH", "MODERATE"))

#For each category of phenotypes, 
# GO term enrichment, plot, and collect top 10
enrich_results = list()
phenotypes = unique(pheno_table$Category)
par(mfrow = c(3, 2))
for (pheno in phenotypes) {
  
  
  myInterestingGenes <- gene_detected %>%
    filter(Category == pheno) %>% 
    pull(Gene) %>% unique()
  geneList <- factor(as.integer(geneNames %in% myInterestingGenes))
  names(geneList) <- geneNames
  str(geneList)
  
  GOdata <- new("topGOdata", ontology = "MF", allGenes = geneList,
   annot = annFUN.gene2GO, gene2GO = geneID2GO)
  
  resultFisher <- runTest(GOdata, algorithm = "classic", statistic = "fisher")
  resultKS <- runTest(GOdata, algorithm = "classic", statistic = "ks")
  resultKS.elim <- runTest(GOdata, algorithm = "elim", statistic = "ks")
  
  #Store results in a table
  allRes <- GenTable(GOdata, classicFisher = resultFisher,
                     classicKS = resultKS, elimKS = resultKS.elim,
                     orderBy = "elimKS", ranksOf = "classicFisher", topNodes = 20)
  enrich_results[[pheno]] =  as_tibble(allRes) %>% mutate(Phenotype = pheno)
  
  #Make automatic plot
  showSigOfNodes(GOdata, score(resultKS.elim), firstSigNodes = 5, useInfo = 'all')
}

bind_rows(enrich_results)

bind_rows(enrich_results) %>%
  dplyr::select(Phenotype, GO.ID, Term, elimKS) %>%
  unite(col = "GO", GO.ID, Term) %>% 
  mutate(Is_enriched = ifelse(elimKS < 0.05, 0, 1)) %>%
  ggplot(aes(y = GO, x = Phenotype, fill = Is_enriched)) +
  geom_tile() +
  theme_bw() +
  theme(axis.title.x = element_blank(), 
        axis.text.x = element_text(angle = 40, hjust = 1, vjust = 1),
        axis.title.y = element_blank(),
        legend.position = "none",
        panel.grid.major.x = element_blank(),
        panel.grid.major.y = element_blank(),
        panel.ontop = TRUE,
        panel.background = element_rect(fill = "transparent")) +
  scale_fill_gradient(low = "white", high = "#82c0cc")
```



## Compare results between phenotypes
We have several variables that could be interpreted the same way. For example, we expect the amount of rain during the driest month and the driest quarter to give us similar information related to drought tolerance. It could be useful to identify useful SNPs to find the ones that are significantly associated to all variables of the same type. Let's see if the p-values seem correlated between variables of the same types and what amount of significant variants are shared.
```{r pairwise_signif}
reshape_GEMMA_results <- function(phenotype_list) {
  temp = results_GEMMA %>%
    filter(Phenotype %in% phenotype_list) %>%
    mutate(logP = -log10(P)) %>%
    filter(logP > 2) %>%
    mutate(Phenotype_code = ifelse(Phenotype == phenotype_list[1], "Pheno1", "Pheno2"))%>%
    dplyr::select(SNP, Phenotype_code, logP, Bonferroni_threshold)
  temp = temp %>% pivot_wider(values_from = logP, names_from = Phenotype_code) %>%
    mutate(Signif_max = ifelse(Pheno1 >= -log10(Bonferroni_threshold), "1", "0"),
           Signif_mean = ifelse(Pheno2 >= -log10(Bonferroni_threshold), "1", "0")) %>%
    unite(Significant, Signif_max, Signif_mean, sep = "", remove = F) 
  
  temp[complete.cases(temp),] 
}

dot_plot_comparison <- function(reshaped_GEMMA_results, phenotype_list) {
  ggplot(temp, aes(x = Pheno1, y = Pheno2, col = Significant)) +
    geom_point(alpha = .7) +
    scale_color_manual(values = c("#E7DFDA", "#489FB5", "#82C0CC", "#FFA62B")) +
    geom_hline(aes(yintercept=-log10(Bonferroni_threshold)),
                 linetype = "dashed", color = "grey20")+
    geom_vline(aes(xintercept=-log10(Bonferroni_threshold)),
                 linetype = "dashed", color = "grey20") +
    theme_classic() +
    labs(x = phenotype_list[1], y = phenotype_list[1])
}



# Variables related to temperatures
# _________________________
#Significant for warm temperatures
warm_temp_var = c("Max Temperature of Warmest Month", "Mean Temperature of Warmest Quarter")
temp = reshape_GEMMA_results(warm_temp_var)
p1 = dot_plot_comparison(temp, warm_temp_var)


#Significant for cold temperatures
cold_temp_var = c("Min Temperature of Coldest Month", "Mean Temperature of Coldest Quarter")
temp = reshape_GEMMA_results(cold_temp_var)
p2 = dot_plot_comparison(temp, cold_temp_var)

row = cowplot::plot_grid(p1 +theme(legend.position = "none"), p2  + theme(legend.position = "none"))
title_gg <- ggplot() + 
  labs(title = str_wrap(paste0("Comparison between the p-values of variables related to temperatures"), 80))
cowplot::plot_grid(title_gg, row, nrow = 2, rel_heights = c(0.1, 1))


# Variables related to rain
# _________________________
drought_var = c("Precipitation of Driest Month", "Precipitation of Driest Quarter")
rainy_var = c("Precipitation of Wettest Month", "Precipitation of Wettest Quarter")
#Significant for drought
temp = reshape_GEMMA_results(drought_var)
p1 = dot_plot_comparison(temp, drought_var)


#Significant for high precipitation
temp = reshape_GEMMA_results(rainy_var)
p2 = dot_plot_comparison(temp, rainy_var)

row = cowplot::plot_grid(p1 +theme(legend.position = "none"), p2  + theme(legend.position = "none"))
title_gg <- ggplot() + 
  labs(title = str_wrap(paste0("Comparison between the p-values of variables related to precipitation"), 80))
cowplot::plot_grid(title_gg, row, nrow = 2, rel_heights = c(0.1, 1))

```

Instead of comparing the results based on a priori knowledge of the underlying wheather/climate patterns for each variables, it is possible to simply look at the correlation between all pairs of variables. This yields a matrix that is similar to the one shown in the previous section but this time it takes into account the p-values in each test and not the values of the bioclimatic variables themselves.
```{r correl climate assoc}
# Keep only variants significant in at least one association test
temp = results_GEMMA %>%
  ungroup() %>%
  dplyr::select(-CHR, -BP, -Nb, -zscore, -Bonferroni_threshold, -Subset, -n) %>%
  distinct() %>%
  pivot_wider(names_from = Phenotype, values_from = P) %>%
  dplyr::select(-SNP) # %>%
  #filter(rowSums(. <= Bonferroni_threshold) >= 1)

#Correlation
Ccor = cor(temp)

#Correlogram
corrplot.mixed(Ccor, upper = "ellipse",
         #col = mycolorsCorrel,
         tl.col="black", tl.cex = 0.5, number.cex = .5)

#Simple heatmap
heatmap(x = Ccor, col = mycolorsCorrel, symm = TRUE)
```

It is also possible that some genes or some variants could be associated to different phenotypes. Here, we can use upset plots to visualize different comparisons:

 * How many genes are found associated between the different phenotypes tested? The genes are identified through the SnpEff pipeline.
 * How many variants are significantly associated to several phenotypes tested?
 * Amongst the variants which are found to be significant to all phenotypes belonging to the same category, how many are found to be common between phenotypes?
 * Amongst the variants which are found to be significant in at least one of the phenotype belonging to the same category, how many are found to be common between phenotypes?
```{r upset}
# Per gene and per phenotype
gene_detected = inner_join(pheno_table, significant) %>%
  mutate(logP = -log10(P)) %>%
  left_join(., signif_ann) %>% 
  mutate(Present = 1) %>%
  ungroup() %>%
  dplyr::select(Phenotype, Gene, Present) %>%
  distinct() %>%
  filter(!is.na(Gene)) %>%
  pivot_wider(names_from = Phenotype, values_from = Present, values_fill = 0)
for_upset = as.data.frame(gene_detected[,-1])
rownames(for_upset) = pull(gene_detected[,1])

upset(for_upset, sets = names(for_upset), 
      order.by = "freq", mb.ratio = c(0.5, 0.5), 
      mainbar.y.label = "Gene numbers")

#Per variant and per phenotype
temp = inner_join(pheno_table, significant) 

temp2 = mutate(temp, Present = 1) %>%
  ungroup() %>%
  dplyr::select(Phenotype, SNP, Present) %>%
  distinct() %>%
  pivot_wider(names_from = Phenotype, values_from = Present, values_fill = 0)
  
for_upset = as.data.frame(temp2[,-1])
rownames(for_upset) = pull(temp2[,1])
upset(for_upset, sets = names(for_upset), 
      order.by = "freq", mb.ratio = c(0.5, 0.5),
      mainbar.y.label = "Variant numbers")

```





## Genomic location of significant variants and genes
Traditionally, GWAS results are represented as Manhattan plot. Let's create some for the main variables we are interested in. 

```{r GEMMA prep plot}
#This table contains the chromosome names (CHR), the cumulative start of the chromosomes (Cumu_start), 
# and the cumulative centers of the chromosomes (Cumu_center). These are used for plotting the chromosomes one
# after the other and to place the chromosome names legends correctly.
chromosomes_lengths = readxl::read_excel(paste0(metadata_dir, "Chromosomes_cumulative_lengths.xlsx")) 


#Prep for simplified function
results_GEMMA = results_GEMMA %>% 
    #inner_join(., chromosomes_lengths)  %>%
    left_join(., significant) %>%
    mutate(SNP_type = ifelse(is.na(SNP_type), "Other", SNP_type)) %>%
    mutate(MAF_type = ifelse(is.na(MAF), "1", ifelse(MAF < 0.05, "1", "19"))) %>%
    mutate(logP = -log10(P))

# Manhattan plot function
Manhattan_one_pheno <- function(results_GEMMA, pheno, color_duo) {
  "To use this function we need to have a table containing the columns CHR, BP,
  logP, Phenotype, SNP_type and MAF_type. CHRs are facets, BP is x, logP is y, SNP_type is color,
  MAF_type is binary shape, and Phenotype is for filtering." 
  temp = results_GEMMA %>% filter(Phenotype == pheno) %>%
    filter(CHR <= 13) %>%
    filter(logP > 2)
  
  ggplot(temp, aes(x=BP, y=logP)) +
    geom_point( aes(color=SNP_type, shape = MAF_type), size=1.3) +
    scale_color_manual(values = color_duo) +
    scale_shape_manual(values =c(1, 19)) +
    geom_hline(aes(yintercept=-log10(Bonferroni_threshold)),
               linetype = "dashed", color = "grey20") +
    theme_bw() +
    theme( legend.position = "none", panel.border = element_blank(),
      panel.grid.major.x = element_blank(), panel.grid.minor.x = element_blank(),
      axis.title.x=element_blank(), axis.text.x=element_blank(),
      axis.ticks.x=element_blank(),
      strip.background = element_rect(colour="white", fill="white"),
      plot.margin = unit(c(0, 0, 0, 0), "cm")) +
      facet_grid(cols = vars(CHR), scales = "free_x") 
}

# Manhattan plot function 2
Manhattan_faceted_pheno <- function(results_GEMMA, pheno_list, color_duo) {
  "To use this function we need to have a table containing the columns CHR, BP,
  logP, Phenotype, SNP_type and MAF_type. CHRs are facets, BP is x, logP is y, SNP_type is color,
  MAF_type is binary shape, and Phenotype is for filtering. We also need a second table 
  found above as chromosomes_lengths." 
  
  results_GEMMA %>% 
    filter(Phenotype %in% pheno_list) %>%
    arrange(CHR, BP) %>% 
    inner_join(., chromosomes_lengths) %>%
    mutate(BPcum = BP + Cumu_start) %>%
    mutate(logP = -log10(P))%>%
    filter(logP > 2) %>%
    ggplot(., aes(x=BPcum, y=logP)) +
      geom_point( aes(color=as.factor(CHR), shape = MAF_type), size=1.3) +
      scale_color_manual(values = rep(color_duo, 22 )) +
    scale_shape_manual(values =c(1, 19)) +
      scale_x_continuous( label = chromosomes_lengths$CHR, breaks= chromosomes_lengths$Cumu_center ) +
      geom_hline(aes(yintercept=-log10(Bonferroni_threshold)),
                 linetype = "dashed", color = "grey20") +
      theme_bw() +
      theme(legend.position="none",
        panel.border = element_blank(),
        panel.grid.major.x = element_blank(),
        panel.grid.minor.x = element_blank()) +
      facet_grid(rows = vars(Phenotype), scales = "free_y") 
}



# Top SNPs
top_SNPs_density <- function(pheno, min_length_locus, custom_colors, min_MAF = 0){
  temp = significant %>% 
    filter(Phenotype == pheno) %>%
    filter(MAF >= min_MAF) %>% 
    group_by(CHR, Locus_nb) %>%
    mutate(Rank = rank(P, ties.method = "first")) %>%
    arrange(Rank) %>%
    ungroup() %>%
    filter(Rank == 1) %>%
    dplyr::select(SNP, CHR, BP)
  
  relevant_climate = climate_per_coordinates %>%
    pivot_longer(-c(X,Y), names_to = "phenotype", values_to = "Bioclim") %>%
    filter(phenotype == pheno) %>%
    left_join(., Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude), 
              by = c("X" = "Longitude", "Y" = "Latitude"))
  
  temp = read_tsv(paste0(GEA_dir, "Significant_GEMMA.ann.tab"), na = ".") %>%
    inner_join(temp, by = c("CHROM" = "CHR", "POS" = "BP")) %>%
    pivot_longer(-c(CHROM, POS, REF, ALT, SNP), 
                 names_to = "ID_file", values_to = "Allele") %>%
    inner_join(., relevant_climate) %>%
    filter(Allele != "NA")
  
  temp %>%
    ggplot(aes(Bioclim, fill = as.factor(Allele), col = as.factor(Allele))) +
    geom_density(alpha = .4) + 
    facet_wrap(vars(CHROM, POS), scales = "free") +
    theme_bw() +
    scale_color_manual(values = custom_colors) +
    scale_fill_manual(values = custom_colors)

}


# Gene/TE tracks plot
plot_gff <- function(gff_name, chromosome, min_coord, max_coord, genes_to_highlight, col = "black") {
  gff = read_tsv(gff_name, 
               col_names = c("CHR", "X1", "mRNA", "Start", "Stop", "X2", "X3", "X4", "Annot"))  %>%
  separate(col = Annot, into = c("ID", "Parent", "Name"), sep = ";") %>%
  mutate(ID = str_remove(ID, "ID=")) %>% 
  filter(CHR == chromosome) %>%
  filter(Start >= min_coord) %>%
  filter(Stop <= max_coord) %>%
  mutate(y_value = rank(Start)) %>%
  arrange(Start)

## If too many genes, organize them
if (nrow(gff) > 5) {
  temp = ceiling(nrow(gff)/5)
  gff = gff %>% mutate(y_value = rep(c(1:5), temp)[1:nrow(gff)])
}

## Make plot
c = gff %>%
  ggplot() +
  geom_segment(mapping = aes(x = Start, xend = Stop, y = y_value, yend = y_value), size = 2, col = col) +
  theme_bw() + 
  theme(axis.title.y = element_blank(), axis.text.y = element_blank(), 
        axis.title.x = element_blank(), axis.ticks.y = element_blank(), 
        panel.grid.minor.y = element_blank(), panel.grid.major.y = element_blank()) +
  coord_cartesian(xlim = c(min_coord, max_coord), ylim = c(min(gff$y_value) - 0.5, max(gff$y_value) + 0.5))

## Highlight interesting genes
temp = match(genes_to_highlight, gff$ID)
if ( sum(!is.na(temp)) > 0 ){
  c = c + geom_segment(data = filter(gff, ID == genes_to_highlight),
                   mapping = aes(x = Start, xend = Stop, y = y_value, yend = y_value), 
                   size = 3, col = "blue")
}
## If there are a reasonable number of genes, 
# I will also add their name on the plot
if (nrow(gff) <= 15) {
  c = c + geom_text(aes(x = Stop + (max_coord - min_coord)*0.05 , y = y_value, label = ID), size = 2) 
}

c
}

```


I want to have a main plot for temperature and for precipitation. First, I choose the maximum and minimum temperature. 
```{r GEMMA plot core temp}

# Manhattan plot with both max and min temperature with mirrored effect
p = Manhattan_one_pheno(results_GEMMA, "Mean Temperature of Warmest Quarter", c("#F5B9B3", "#F28482")) +
  labs(y = str_wrap("Mean Temperature of Warmest Quarter (-logP)", 25))

q = Manhattan_one_pheno(results_GEMMA, "Min Temperature of Coldest Month", c("#C4E7FD", "#0889D9")) +
  labs(y = str_wrap("Min Temperature of Coldest Month (-logP)", 25)) +
  scale_y_reverse() + theme(strip.text.x = element_blank())

cowplot::plot_grid(p, q, nrow = 2)

#ggsave(paste0(fig_dir, "GEA_Manhattan_temperature_main.pdf"), width = 18, height = 10, units = "cm")

top_SNPs_density("Mean Temperature of Warmest Quarter", 2000, c("#F5B9B3", "#F28482", "grey"), 0.05)
top_SNPs_density("Min Temperature of Coldest Month", 2000, c("#C4E7FD", "#0889D9", "grey"), 0.05)
```


Then, for precipitation, I choose the precipitation of the driest and of the wettest month.
```{r GEMMA core rain}
#  Manhattan plot for core precipitation
p = Manhattan_one_pheno(results_GEMMA, "Precipitation of Driest Month", c("#FFD399", "#ff9f1c")) +
  labs(y = str_wrap("Precipitation of Driest Month (-logP)", 25))

q = Manhattan_one_pheno(results_GEMMA, "Precipitation of Wettest Month", c("#cbf3f0", "#2ec4b6")) +
  scale_y_reverse() + theme(strip.text.x = element_blank()) +
  labs(y = str_wrap("Precipitation of Wettest Month (-logP)", 25))

cowplot::plot_grid(p, q, nrow = 2)

ggsave(paste0(fig_dir, "GEA_Manhattan_rain_main.pdf"), width = 18, height = 10, units = "cm")

top_SNPs_density("Precipitation of Driest Month", 1000, c("#FFD399", "#ff9f1c", "grey"), 0.05)
top_SNPs_density("Precipitation of Wettest Month", 1000, c("#cbf3f0", "#2ec4b6", "grey"), 0.05)
```


I also want to have a Manhattan plot for all possible environmental variable. I do try to classify them, grouping conceptually similar variables together.
```{r GEMMA plots}

#Yearly variables
Manhattan_faceted_pheno(results_GEMMA, Bioclim_var[c(2, 3, 7)], c("grey", "#82c0cc")) +
  labs(title = "Ranges bioclimatic variables")
Manhattan_faceted_pheno(results_GEMMA, Bioclim_var[c(4, 15)], c("grey", "#82c0cc")) +
  labs(title = "Seasonality bioclimatic variables")
Manhattan_faceted_pheno(results_GEMMA, Bioclim_var[c(1, 12)], c("grey", "#82c0cc")) +
  labs(title = "Annual rain and temperature")


#Rain
Manhattan_faceted_pheno(results_GEMMA, drought_var, c("grey", "#82c0cc")) +
  labs(title = "Drought")
Manhattan_faceted_pheno(results_GEMMA, rainy_var, c("grey", "#82c0cc")) +
  labs(title = "Rain")

#Mix rain and temperature
Manhattan_faceted_pheno(results_GEMMA, Bioclim_var[c(8, 9, 18, 19)], c("grey", "#82c0cc")) +
  labs(title = "Mix rain and temperature")


#Temperatures
p = Manhattan_faceted_pheno(results_GEMMA, warm_temp_var, c("grey", "#82c0cc")) +
  labs(title = "High temperatures")
q = Manhattan_faceted_pheno(results_GEMMA, cold_temp_var, c("grey", "#82c0cc")) +
  labs(title = "Cold temperatures")
p
q
```

```{r QTLs}
## Adding QTL data from Lendenman
QTL = readxl::read_excel(path = paste0(GEA_dir, "Lendenman_QTL.xlsx")) %>%
  inner_join(chromosomes_lengths) %>%
  mutate( start = 1000*start, end = 1000*end, peak = 1000*peak - distance_threshold, peak2 = peak + distance_threshold,
          Start = start + Cumu_start, End = end + Cumu_start, Peak = peak + Cumu_start) %>%
  rename(Observed_pheno = Phenotype)

temp = bind_rows(mutate(QTL, Phenotype = warm_temp_var[[1]]),
                  mutate(QTL, Phenotype = warm_temp_var[[2]]))
p + geom_segment(data = temp, 
                aes(x = Peak, xend = Peak, y = 0.5, yend = 1.6 ),
                arrow = arrow(length = unit(0.03, "npc")))

temp = bind_rows(mutate(QTL, Phenotype = cold_temp_var[[1]]),
                  mutate(QTL, Phenotype = cold_temp_var[[2]]))
q + geom_segment(data = temp, 
                aes(x = Peak, xend = Peak, y = 0.5, yend = 1.6),
                arrow = arrow(length = unit(0.03, "npc")), col = "grey40")


QTL_GR = makeGRangesFromDataFrame(dplyr::select(QTL, chr = CHR, start= peak, end =peak2))
temp = significant %>% 
  group_by(CHR, Locus_nb, Phenotype, score = Nb) %>% 
  summarize(start = min(BP) - distance_threshold, end = max(BP) + distance_threshold) %>% 
  inner_join(filter(pheno_table, Category == "Temperature")) %>%
  mutate(strand = ".") %>%
  dplyr::select(CHR, start, end) 
signif_GR = makeGRangesFromDataFrame(temp)
as.tibble(GenomicRanges::intersect(QTL_GR, signif_GR))
```

It is clear from the Manhattan plots that some significant loci are shared between phenotypes, especially phenotypes that we would intuitively group together such as the maximum temperature and the average temperature of the three warmest months. 


```{r Heatmaps}
#pheno_category = "Temperature"
#top_SNPs_density("Precipitation of Driest Month", 2000, c("#FFD399", "#ff9f1c", "grey"))
#top_SNPs_density("Precipitation of Wettest Month", 2000, c("#cbf3f0", "#2ec4b6", "grey"))
#top_SNPs_density("Mean Temperature of Warmest Quarter", 2000, c("#F5B9B3", "#F28482", "grey"), 0.05)
#top_SNPs_density("Min Temperature of Coldest Month", 2000, c("#C4E7FD", "#0889D9", "grey"), 0.05)


for (pheno_category in c("Temperature", "Precipitation")){
  
  #Define new, combined loci per phenotypic category
  relevant_positions = full_join(significant, pheno_table) %>% 
    filter(Category == pheno_category)
  new_loci = from_positions_to_loci(relevant_positions, distance_threshold)
  relevant_positions = relevant_positions %>%
    dplyr::select(-Locus_nb) %>%
    full_join(., new_loci) %>%
    distinct()
  locus_signif_in = relevant_positions %>% 
    dplyr::select(Category, Sub_category, Phenotype, CHR, Locus_nb) %>% 
    distinct() %>% ungroup()
  sum_locus_signif_in = locus_signif_in %>% group_by(CHR, Locus_nb) %>% dplyr::count() %>%
    rename(number_of_biovar_per_locus = n)
  
  #Finding position with the lowest p-value per locus
  temp = relevant_positions %>%
    group_by(CHR, Locus_nb, SNP) %>%
    mutate(number_of_biovar_per_SNP = n()) %>%
    left_join(sum_locus_signif_in) %>%
    filter((number_of_biovar_per_locus - number_of_biovar_per_SNP) <= 1) %>%
    group_by(CHR, Locus_nb) %>%
    mutate(Rank = rank(P, ties.method = "first")) %>%
    arrange(Rank) %>%
    ungroup() %>%
    filter(Rank == 1) %>%
    dplyr::select(SNP, CHROM = CHR, POS = BP, Category, Sub_category, Phenotype) 
  
  #Adding isolate alleles information
  temp = temp %>%
    inner_join(., read_tsv(paste0(GEA_dir, "Significant_GEMMA.ann.tab"), na = ".")) %>%
    pivot_longer(-c(CHROM, POS, REF, ALT, SNP, Category, Sub_category, Phenotype), 
                   names_to = "ID_file", values_to = "Allele") %>%
    inner_join(., Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude, Continent)) %>%
    mutate(Latitude = round(Latitude/2)*2, Longitude = round(Longitude/2)*2) 

  
  #To identify the allele associated with high and with low values of the bioclimatic variable
  med_bioclim_per_allele = climate_per_coordinates %>%
      pivot_longer(-c(X,Y), names_to = "Phenotype", values_to = "Bioclim") %>%
      left_join(., pheno_table)%>%
      left_join(., Zt_meta  %>% dplyr::select(ID_file, Latitude, Longitude, Continent), 
                by = c("X" = "Longitude", "Y" = "Latitude")) %>%
      full_join(., temp) %>%
      filter(!is.na(Allele) & !is.na(Phenotype))  %>% 
      filter(Category == pheno_category) %>%
      group_by(Category, Phenotype, SNP, Allele) %>%
      dplyr::summarize(Median_bioclim = median(Bioclim, na.rm = T))  %>%
      mutate(Phenotype_hilow = ifelse(Median_bioclim == max(Median_bioclim), "High", "Low"))
  
  #From the allelic data, get freq per place with at least 5 isolates
  temp3 = ungroup(inner_join(temp, med_bioclim_per_allele)) %>% 
    dplyr::select(Phenotype, ID_file, SNP, Longitude, Latitude, Allele, Continent, Phenotype_hilow) %>%
    filter(!is.na(Allele))%>%
    group_by(Phenotype, SNP, Continent, Longitude, Latitude, Phenotype_hilow) %>%
    dplyr::count(Phenotype_hilow)  %>%
    pivot_wider(names_from = Phenotype_hilow, values_from = n, values_fill = 0) %>%
    mutate(sum = High + Low,
           High = High/sum) %>%
    filter(sum >= 5) %>%
    unite(col = "Place", c(Continent, Longitude, Latitude))%>%
    dplyr::select(Place, SNP, High) 
  
  #Select only variants which are found in high freq
  # in at least one place
  SNP_to_plot = temp3 %>% 
    group_by(SNP) %>% 
    summarize(minimum = min(High), maximum = max(High)) %>%
    filter(maximum >= 0.7) %>%
    filter(minimum <= 0.3) %>%
    dplyr::select(-maximum, -minimum)
  
  #Widen the table and remove duplicate variants
  temp3 = inner_join(temp3, SNP_to_plot)%>%
    pivot_wider(names_from = Place, values_from = High, values_fill = 0)%>%
    ungroup() %>%
    dplyr::select(-Phenotype) %>%
    distinct()
  
  #Get annotations for the places
  # - First, continent
  # - TODO, genetic group
  temp = as_tibble(colnames(temp3[,-1])) %>%
    separate(value, into = c("Continent", "X", "Y"), sep = "_") %>%
    dplyr::select(Continent) %>% pull()
  annotation_col <- data.frame(
          Continent = temp)
  rownames(annotation_col) = colnames(temp3[,-1])
  ## Define colors
  temp2 = c("#DA4167", "#ffba0a", "#A20106", "#3F6E0C", "#129eba", "#8fa253" )
  names(temp2) = sort(unique(temp))
  colors_pheatmap = list(Continent = temp2)
  
  #Get annot for the SNPs
  temp = dplyr::select(relevant_positions, SNP, CHR, Locus_nb) %>% 
    left_join(locus_signif_in) %>% 
    dplyr::select(SNP, Sub_category) %>% distinct() %>% 
    mutate(Is_signif = "Significant") %>% 
    pivot_wider(names_from = Sub_category, values_from = Is_signif, values_fill = "Non-significant") %>%
    filter(!is.na(SNP)) %>% 
    filter(SNP %in% pull(SNP_to_plot)) 
  annotation_row <- data.frame(Annual = temp["Annual"], Low = temp["Low"], 
                           High= temp["High"])
  rownames(annotation_row) = pull(temp["SNP"])
  ## Define colors
  temp2 = c("White", "Grey")
  names(temp2) = c("Non-significant", "Significant")
  colors_pheatmap$High = temp2
  colors_pheatmap$Low = temp2
  colors_pheatmap$Annual = temp2
  
  #Define relevant colors for the heatmap itself.
  if (pheno_category == "Temperature"){
      color_heatmap = colorRampPalette(c("#0889D9","white","#F28482"))(30)
  } else if (pheno_category == "Precipitation") {
      color_heatmap = colorRampPalette(c("#ff9f1c","white","#2ec4b6"))(30)
  } else {
      color_heatmap = colorRampPalette(c("#grey","white","#82c0cc"))(30)
  }
  
  
  #Make the heatmap
  for_pheatmap = as.matrix(temp3[,-1])
  rownames(for_pheatmap) = pull(temp3[,1])
  pheatmap(for_pheatmap, 
           angle_col = "45", 
           fontsize = 8, 
           color = color_heatmap, border_color = NA,
           cluster_cols = F, #cluster_rows = F,
           annotation_col = annotation_col,
           annotation_colors = colors_pheatmap,
           annotation_row = annotation_row,
           show_rownames = F)

}
```


### Genes per chromosomes
```{r genes MAF 5}

#Translating SnpEff effects in a numerical scale
gene_detected = inner_join(pheno_table, significant) %>%
    mutate(logP = -log10(P)) %>%
    left_join(., signif_ann) %>% 
  mutate(Ranked_effect_strength = Effect_strength)
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "HIGH"] = 1
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODERATE"] = 2
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "LOW"] = 3
gene_detected$Ranked_effect_strength[gene_detected$Ranked_effect_strength == "MODIFIER"] = 4
gene_detected$Ranked_effect_strength[is.na(gene_detected$Effect_strength)] = 5
gene_detected$Effect_strength[is.na(gene_detected$Effect_strength)] = "NO_EFFECT"



plot_tiles_GWAS <- function(list_chr, list_pheno_cat) {
  if (list_pheno_cat == "all"){
    temp = gene_detected %>% 
      filter( MAF > 0.05 ) %>%
      filter(CHR %in% list_chr)  
  } else {
      temp = gene_detected %>% 
        filter( MAF > 0.05 ) %>%
        filter(CHR %in% list_chr) %>%
        filter(Category %in% list_pheno_cat) 
  }
  
  ggplot(temp, aes(x = Gene, y = Phenotype, fill = log(P))) + 
    geom_tile() + 
    theme_bw() +
    theme(axis.title.x = element_blank(), 
          axis.text.x = element_text(angle = 40, hjust = 1, vjust = 1),
          axis.title.y = element_blank(),
          legend.position = "none",
          panel.grid.major.x = element_blank(),
          panel.grid.major.y = element_blank(),
          panel.ontop = TRUE,
          panel.background = element_rect(fill = "transparent")) +
    scale_fill_gradient(low = "white", high = "#82c0cc")
}

gene_detected %>% 
  filter( MAF > 0.05 ) %>%
  dplyr::select(CHR, Gene, Phenotype) %>%
  distinct() %>%
  group_by(CHR) %>%
  dplyr::count() %>%
  arrange(n)

#c("Temperature", "Precipitation")
plot_tiles_GWAS(c(8,9,3,4,10), "all") +
    labs(title = "Five chromosomes with the lowest number of hits") +
  facet_grid(rows = vars(Category), scales = "free")
    

plot_tiles_GWAS(c(5, 11, 2, 6), c("Temperature", "Precipitation")) +
    facet_grid(rows = vars(Category), cols = vars(CHR), scales = "free")

plot_tiles_GWAS(c(13), c("Temperature", "Precipitation")) +
    labs(title = "Chromosome 13")+
  facet_grid(rows = vars(Category), scales = "free")

plot_tiles_GWAS(c(1), c("Temperature", "Precipitation"))+
    labs(title = "Chromosome 1")+
  facet_grid(rows = vars(Category), scales = "free")

plot_tiles_GWAS(c(7), c("Temperature", "Precipitation")) +
    labs(title = "Chromosome 7")+
  facet_grid(rows = vars(Category), scales = "free")



```



### Manhattan plots and genes in specific loci

```{r map allele}
i=10
#SNP_name = "3_444010"
#SNP_name = "7_1449518"
SNP_name = "10_460207"
temp2 = raster(paste0(data_dir,"Climatic_data/wc2.1_10m_bio/wc2.1_10m_bio_", i, ".tif"))
temp = as.data.frame(rasterToPoints(temp2, xy = TRUE))
colnames(temp) <- c("X", "Y", "Bioclim_var")

map1 = ggplot() +
  geom_raster(data = temp , 
              aes(x = X, y = Y, 
                  fill = Bioclim_var)) + 
  theme_void() +
  scale_fill_viridis("inferno") +
  theme(panel.border  = element_rect(fill=NA, colour = "grey",size = 0.5, linetype = 1))
#panel.background = element_rect(fill = "aliceblue"))
p1 = map1 + coord_cartesian(xlim=c(-20, 60), ylim=c(20, 70))
p2 = map1 + coord_cartesian(xlim=c(-160, -40), ylim=c(-80, 80))
p3 = map1 + coord_cartesian(xlim=c(105, 175), ylim=c(-65, 10))
aus_map = cowplot::plot_grid(p3 + theme(legend.position = "None"), get_legend(p1), ncol = 2, rel_widths = c(1.1, 1))
ligne = cowplot::plot_grid(p1 + theme(legend.position = "None"), aus_map, ncol = 1,  rel_heights = c(1, 0.7))
cowplot::plot_grid(p2 + theme(legend.position = "None"), ligne, ncol = 2, rel_widths = c(0.7, 1)) 
ggsave(paste0(fig_dir, "Map_BIO", i, ".pdf"), width = 18, height = 10, units = "cm")

allele = significant %>% 
    filter(Phenotype == Bioclim_var[i]) %>% 
    filter(SNP == SNP_name) %>%
    dplyr::select(SNP, CHR, BP) %>%
    inner_join(., read_tsv(paste0(GEA_dir, "Significant_GEMMA.ann.tab"), na = "."), by = c("CHR" = "CHROM", "BP" = "POS")) %>%
    pivot_longer(-c(CHR, BP, REF, ALT, SNP), 
                 names_to = "ID_file", values_to = "Allele") %>%
    inner_join(., Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude)) %>%
    mutate(Latitude = round(Latitude/2)*2, Longitude = round(Longitude/2)*2)%>%
    mutate(Allele = ifelse(Allele == 0, "REF", ifelse(Allele == 1, "ALT1", "ALT2"))) %>%
    group_by(Allele, Longitude, Latitude) %>%
    dplyr::count() %>%
    pivot_wider(names_from = Allele, values_from = n, values_fill = 0) %>%
    filter(!is.na(Longitude)) %>%
    mutate(Radius = 2) 

#Add if column alt2
if ("ALT2" %in% colnames(allele)) {
  allele = allele
} else {
  allele = allele %>% mutate(ALT2 = 0)
}

p1 = ggplot() + coord_cartesian(xlim=c(-20, 60), ylim=c(20, 70)) + 
  geom_scatterpie(data = allele, mapping = aes(x=Longitude, y=Latitude, r = Radius), 
                  cols = c("REF", "ALT1", "ALT2"), color=NA, alpha=.8) +
  theme(panel.background = element_rect(fill = "transparent"), # bg of the panel
    plot.background = element_rect(fill = "transparent", color = NA), # bg of the plot
    panel.grid.major = element_blank(), # get rid of major grid
    panel.grid.minor = element_blank(), # get rid of minor grid
    legend.background = element_rect(fill = "transparent"), # get rid of legend bg
    legend.box.background = element_rect(fill = "transparent"))

p2 = ggplot() + coord_cartesian(xlim=c(-160, -40), ylim=c(-80, 80)) + 
  geom_scatterpie(data = allele, mapping = aes(x=Longitude, y=Latitude, r = Radius*2), 
                  cols = c("REF", "ALT1", "ALT2"), color=NA, alpha=.8)+
  theme(panel.background = element_rect(fill = "transparent"), # bg of the panel
    plot.background = element_rect(fill = "transparent", color = NA), # bg of the plot
    panel.grid.major = element_blank(), # get rid of major grid
    panel.grid.minor = element_blank(), # get rid of minor grid
    legend.background = element_rect(fill = "transparent"), # get rid of legend bg
    legend.box.background = element_rect(fill = "transparent"))

p3 = ggplot() + coord_cartesian(xlim=c(105, 175), ylim=c(-65, 10)) + 
  geom_scatterpie(data = allele, mapping = aes(x=Longitude, y=Latitude, r = Radius*2), 
                  cols = c("REF", "ALT1", "ALT2"), color=NA, alpha=.8)+
  theme(panel.background = element_rect(fill = "transparent"), # bg of the panel
    plot.background = element_rect(fill = "transparent", color = NA), # bg of the plot
    panel.grid.major = element_blank(), # get rid of major grid
    panel.grid.minor = element_blank(), # get rid of minor grid
    legend.background = element_rect(fill = "transparent"), # get rid of legend bg
    legend.box.background = element_rect(fill = "transparent"))

#Plotting all the maps together! 
aus_map = cowplot::plot_grid(p3 + theme(legend.position = "None"), get_legend(p1), ncol = 2, rel_widths = c(1.1, 1))
ligne = cowplot::plot_grid(p1 + theme(legend.position = "None"), aus_map, ncol = 1,  rel_heights = c(1, 0.7))
cowplot::plot_grid(p2 + theme(legend.position = "None"), ligne, ncol = 2, rel_widths = c(0.7, 1)) 
ggsave(paste0(fig_dir, "Allele_", SNP_name,"_BIO", i, ".pdf"), width = 18, height = 10, units = "cm")

```

```{r plot GWAS locus}


#Set parameters for the plot
chromosome = 1
min_coord = 2076000
max_coord = 2125000

#chromosome = 1
#min_coord = 2020000
#max_coord = 2040000
#pheno = "Precipitation of Driest Month"
pheno = "Precipitation of Coldest Quarter"
#pheno = "Max Temperature of Warmest Month"

#pheno = "Precipitation of Warmest Quarter"
genes_to_highlight = c("Zt09_model_5_00004")
TE_gff = "~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Cecile_IPO323_refTEs_match.gff"
gene_gff = "~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.Grandaubert2015.mRNA.gff3"

# Prepare the gene section of the plot
c = plot_gff(TE_gff, chromosome, min_coord, max_coord, genes_to_highlight, "grey50")
d = plot_gff(gene_gff, chromosome, min_coord, max_coord, genes_to_highlight)

# Prepare the GWAS section of the plot for the first phenotype
b = results_GEMMA %>%
  filter(Phenotype == pheno) %>%
  filter(CHR == chromosome) %>%
  filter(BP >= min_coord) %>%
  filter(BP <= max_coord) %>%
  mutate(logP = -log10(P)) %>%
  ggplot() +
  geom_point(aes(x = BP, y = logP, col = logP)) +
    geom_hline(aes(yintercept=-log10(Bonferroni_threshold)),
               linetype = "dashed", color = "grey20") +
  theme_bw() +
  labs(title = paste0("GWAS: ", pheno),
       subtitle = paste0("The region is on the chromosome ", chromosome, 
                           " from ", min_coord, "bp to ", max_coord, "bp.")) +
  theme(axis.title.x = element_blank(),
        panel.grid.major =  element_line(color = "grey90"),
        axis.text = element_text(color = "grey20", size = 9))+
  coord_cartesian(xlim = c(min_coord, max_coord))

cowplot::plot_grid(b, c, d, ncol = 1, align = 'v', axis = 'lr', rel_heights = c(1, 0.3, 0.3))

```











# GEA: without extreme climatic values
***

```{r filter extreme BIO}

#Establishing summary stats and thresholds
sum_stat_climate= climate_per_coordinates %>%
  pivot_longer(cols = -c(X, Y), names_to = "Phenotype", values_to = "Value") %>%
  group_by(Phenotype) %>%
  summarize(median = median(Value), stdvar = sd(Value))

climate_per_coordinates %>%
  pivot_longer(cols = -c(X, Y), names_to = "Phenotype", values_to = "Value") %>%
  ggplot(aes(Value)) +
  geom_density(fill = "#EDE7E3", col ="#ADA7A9") +
  facet_wrap(.~Phenotype, scales = "free") +
  theme_classic() + 
  geom_vline(data = sum_stat_climate, aes(xintercept = median), col = "#82c0cc") + 
  geom_vline(data = sum_stat_climate, aes(xintercept = median - 2*stdvar), linetype="dashed", col = "#82c0cc") + 
  geom_vline(data = sum_stat_climate, aes(xintercept = median + 2*stdvar), linetype="dashed", col = "#82c0cc") +
  labs(title = "Distribution of climatic variables values",
       subtitle = str_wrap(paste0("The line represents de median and the dashed lines",
                                  "the thresholds of median +/- 2sd. "),
                           width = 100))

#Determining who should be filtered or not
filtering_extreme_climate = climate_per_coordinates %>%
  pivot_longer(cols = -c(X, Y), names_to = "Phenotype", values_to = "Value") %>%
  full_join(., sum_stat_climate) %>%
  full_join(Zt_meta %>% dplyr::select(ID_file, Latitude, Longitude), .,
                   by = c("Latitude" = "Y", "Longitude" = "X"))  %>% 
  mutate(to_filter = case_when(
  Value > (median + 2*stdvar) ~ "Filter_out",
  Value < (median - 2*stdvar) ~ "Filter_out"), 
  to_filter = ifelse(is.na(to_filter), "Keep", to_filter))


#Graphical summary of proportions of kept/filtered samples 
temp = dplyr::count(filtering_extreme_climate, to_filter, Phenotype)

ggplot(data = filter(temp, !is.na(Phenotype)), aes(x = Phenotype, y = n, fill = to_filter)) +
  geom_bar(stat = "identity") + 
  theme_bw() +
  coord_flip() +
  geom_text(data = temp %>% filter(to_filter == "Keep") %>%
              filter(!is.na(Phenotype)),
            aes(label = paste0(n, "/1109"), y = n/2), col = "white") + 
  scale_fill_manual(values = c("#EDE7E3", "#82c0cc")) +
  theme(panel.grid.major.y = element_blank(), panel.grid.minor.y = element_blank(), 
        axis.title.y = element_blank()) +
  labs(y = "Number of genomes", fill = "Filtered status",
       title = "Filtering of extreme climatic values",
       subtitle = str_wrap("The isolates displaying a value above/below the median +/- 2*sd are filtered out.",
                           width = 60))

#Write subset files
#for (bio_var in c(5, 10, 6, 11, 13, 16, 14, 17)){
for (bio_var in c(1,2,3,4,7,8,9,12,15,18,19)){
  temp = inner_join(filtering_extreme_climate, pheno_table) %>%
    filter(BIO_ID == paste0("BIO", bio_var)) %>%
    left_join(read_tsv(Zt_list, col_names = "ID_file") %>% mutate(ID2 = ID_file),.) %>%
    filter(to_filter == "Keep") %>%
    dplyr::select(ID_file, ID2, Value) 
  bind_cols(temp, tibble(X1 = rep(0, nrow(temp)), X2 = rep(0, nrow(temp)), X3 = rep(0, nrow(temp)))) %>%
    dplyr::select(ID_file, ID2, X1, X2, X3, Value) %>%
    write_tsv(path = paste0(GEA_dir, "Phenotypes_with_climatic_variables_BIO", bio_var, ".fam"), col_names = F)
}
#Transfer on the cluster
#rsync -avP ~/Documents/Postdoc_Bruce/Projects/WW_project/5_GEA/Phenotypes_with_climatic_variables_BIO* alice@130.125.25.244:/data2/alice/WW_project/5_GEA/

```


```{r GEMMAres wo extr}
subset_BIO = c(5, 6, 10, 11, 13, 14, 16, 14, 17)

results_GEMMA_wo_extr = list()

for (i in subset_BIO) {
  bio_ID = paste0("BIO", i)
  phenotype = filter(pheno_table, BIO_ID == bio_ID) %>% pull(Phenotype)

  temp = list.files(path = paste0(GEA_dir, "2_Subsets/"), pattern = paste0("*subset_", bio_ID,".assoc.txt")) %>% 
    map_df(~read_tsv(file = paste0(GEA_dir, "2_Subsets/", .), col_types = c("dcddccdddddd"))) %>%
    unite(col = SNP, chr, ps, sep = "_", remove = F) %>%
    dplyr::select(SNP, CHR = chr, BP = ps, P = p_wald, zscore = logl_H1) %>%
    mutate(Phenotype = phenotype, Nb = i, )
  results_GEMMA_wo_extr[[phenotype]] = temp
}

results_GEMMA_wo_extr = bind_rows(results_GEMMA_wo_extr)


#Genomic Inflation Factor
temp = results_GEMMA_wo_extr %>% 
        group_by(Phenotype) %>%
        summarize(Genomic_Inflation_Factor = median(qchisq(1-P,1))/qchisq(0.5,1))
BIO_sum_table =  inner_join(pheno_table, temp) %>% arrange(Genomic_Inflation_Factor)       
kable(BIO_sum_table)

#QQ-plots
par(mfrow=c(3,3)) 
for (i in subset_BIO){
  phenotype = filter(pheno_table, BIO_ID == bio_ID) %>% pull(Phenotype)
temp = results_GEMMA_wo_extr %>% 
  filter(Phenotype == phenotype) %>% 
  dplyr::sample_frac(size = .10) %>% 
  dplyr::select(P) %>% pull()
GWASTools::qqPlot(temp, thinThreshold = 2)
}



```

```{r}
head(results_GEMMA_wo_extr) 
Bonferroni_threshold_wo_extr = results_GEMMA_wo_extr %>% 
  group_by(Phenotype, Nb) %>%
  dplyr::count() %>%
  mutate(Bonferroni_threshold = 0.05/n)

results_GEMMA_wo_extr = full_join(results_GEMMA_wo_extr, Bonferroni_threshold_wo_extr)

significant_wo_extr = results_GEMMA_wo_extr  %>%
  filter(P <= Bonferroni_threshold)  %>% group_by(SNP, CHR, BP) %>%
  mutate(Count = n()) %>% mutate(SNP_type = "significant") %>% 
  ungroup() 

temp = results_GEMMA_wo_extr %>% 
  filter(P <= Bonferroni_threshold) %>% 
  dplyr::count(Phenotype) %>% 
  arrange(n) %>%
  dplyr::select(Phenotype, Nb_signif_variants_wo_extr = n)

temp2 = results_GEMMA %>% 
  filter(P <= Bonferroni_threshold) %>% 
  dplyr::count(Phenotype) %>% 
  arrange(n) %>%
  dplyr::select(Phenotype, Nb_signif_variants = n)

inner_join(temp, temp2)
temp = bind_rows(results_GEMMA_wo_extr %>% mutate(Isolate_group = "No_extreme"),
          results_GEMMA %>% mutate(Isolate_group = "ALL")) %>%
  filter(Nb %in% subset_BIO) %>%
  mutate(logP =  -log10(P)) %>%
  filter(logP > 2) %>%
  dplyr::select(Isolate_group, SNP, logP, Phenotype, Nb) %>%
  pivot_wider(names_from = Isolate_group, values_from = logP) %>%
  filter(!is.na(ALL)) %>%
  filter(!is.na(No_extreme))

ggplot(temp, aes(x = ALL, y = No_extreme)) +
  geom_point(alpha =.4) + 
  facet_wrap(vars(Phenotype), scales = "free") +
  theme_bw() +
  geom_abline(slope = 1, intercept = 0)

```
```{r}
#Prep for simplified function
results_GEMMA_wo_extr = results_GEMMA_wo_extr %>% 
    #inner_join(., chromosomes_lengths)  %>%
    left_join(., significant_wo_extr %>% dplyr::filter(Phenotype == pheno) ) %>%
    mutate(SNP_type = ifelse(is.na(SNP_type), "Other", SNP_type)) %>%
    mutate(MAF_type = "1") %>%
    #mutate(MAF_type = ifelse(is.na(MAF), "1", ifelse(MAF < 0.05, "1", "19"))) %>%
    mutate(logP = -log10(P))

p = Manhattan_one_pheno(results_GEMMA_wo_extr, "Mean Temperature of Warmest Quarter", c("#F5B9B3", "#F28482")) +
  labs(y = str_wrap("Mean Temperature of Warmest Quarter (-logP)", 25))

q = Manhattan_one_pheno(results_GEMMA_wo_extr, "Min Temperature of Coldest Month", c("#C4E7FD", "#0889D9")) +
  labs(y = str_wrap("Min Temperature of Coldest Month (-logP)", 25)) +
  scale_y_reverse() + theme(strip.text.x = element_blank())

cowplot::plot_grid(p, q, nrow = 2)

Manhattan_faceted_pheno(results_GEMMA, drought_var, c("grey", "#82c0cc")) +
  labs(title = "Drought")
Manhattan_faceted_pheno(results_GEMMA_wo_extr, drought_var, c("grey", "#f5cac3")) +
  labs(title = "Drought")
Manhattan_faceted_pheno(results_GEMMA, rainy_var, c("grey", "#82c0cc")) +
  labs(title = "Rain")
Manhattan_faceted_pheno(results_GEMMA_wo_extr, rainy_var, c("grey", "#f5cac3")) +
  labs(title = "Rain")


Manhattan_faceted_pheno(results_GEMMA, warm_temp_var, c("grey", "#82c0cc")) +
  labs(title = "High temperatures")
Manhattan_faceted_pheno(results_GEMMA_wo_extr, warm_temp_var, c("grey", "#f5cac3")) +
  labs(title = "High temperatures")
Manhattan_faceted_pheno(results_GEMMA, cold_temp_var, c("grey", "#82c0cc")) +
  labs(title = "Cold temperatures")
Manhattan_faceted_pheno(results_GEMMA_wo_extr, cold_temp_var, c("grey", "#f5cac3")) +
  labs(title = "Cold temperatures")
```





# Fungicide resistance
***

This is based on the same scripts made for Fanny Hartmann's paper on the pairs of populations: looking for alleles already known to be involved in resistance to fungicide.

```{r bar plot fung}
genotypes_resistance = read_tsv(paste0(fung_dir, "genotypes_tidy_format.tab"),
                                col_names = c("temp", "ID_file", "Allele")) %>%
  separate(temp, into = c("Gene", "AA_change"), sep = "\\.") %>%
   dplyr::mutate(AA_REF = str_sub(AA_change, 1, 1)) %>%
  dplyr::mutate(Allele_type = ifelse(AA_REF == Allele, "Origin", "Mutant")) %>%
  left_join(Zt_meta)


genotypes_resistance %>%
  dplyr::count(AA_change, Gene, Allele_type) %>%
  ggplot(aes(x=n, y=AA_change, fill = Allele_type)) +
  geom_barh(stat="identity") +
  facet_grid(Gene ~ ., scales = "free_y", space = "free_y")



```

I suspect that there will be an effect of both geography and time on the frequency of these alleles and so I try to represent this here.


```{r fungicides bubble plots}
#CYP51
temp = genotypes_resistance %>%
  filter(Gene == "CYP51") %>%
  dplyr::count(AA_change, Continent,Sampling_Date, Allele_type) %>%
  pivot_wider(names_from = Allele_type, values_from = n, values_fill = list(n = 0)) %>%
  mutate(prop_mutant = Mutant/(Mutant + Origin)) %>%
  filter(!is.na(Sampling_Date)) %>%
  filter(!is.na(Continent)) %>%
  group_by(AA_change) %>%
  dplyr::mutate(Somme = sum(Mutant)) %>%
  filter(Somme > 0)

temp %>%
  ggplot(aes(x=Sampling_Date, y=Continent, size = prop_mutant, fill = prop_mutant)) +
  geom_point(shape=21, alpha =0.7) + theme_bw() +
  facet_wrap(.~AA_change) + 
  scale_fill_gradient(low="white", high = "#16697a") +
  labs(title = str_wrap(paste("Mutant proportion per known mutation",
                              "in the gene CYP51"), width = 35),
       subtitle = str_wrap(paste("This gene can mutate and cause resistance to azole fungicides."), width = 65),
       fill = "Proportion of mutants",
       size = "Proportion of mutants",
       x = "Years", y = "")

#Beta_tubulin
genotypes_resistance %>%
  filter(Gene == "beta_tubulin") %>%
  dplyr::count(AA_change, Continent,Sampling_Date, Allele_type) %>%
  pivot_wider(names_from = Allele_type, values_from = n, values_fill = list(n = 0)) %>%
  mutate(Number_all = Mutant + Origin) %>%
  mutate(prop_mutant = Mutant/Number_all) %>%
  group_by(AA_change) %>%
  dplyr::mutate(Somme = sum(Mutant)) %>%
  filter(Somme > 0) %>%
  ggplot(aes(x=Sampling_Date, y=Continent, size = prop_mutant, fill = prop_mutant)) +
  geom_point(shape=21, alpha =0.7) + theme_bw() +
  facet_wrap(.~AA_change) + scale_fill_gradient(low="white", high = "#16697a")+
  labs(title = str_wrap(paste("Mutant proportion per known mutation",
                              "in the beta tubuline gene"), width = 35),
       subtitle = str_wrap(paste("This gene can mutate and cause resistance",
                                 " to benzimidazole fungicides."), width = 65),
       fill = "Proportion of mutants",
       size = "Proportion of mutants",
       x = "Years", y = "")

# Mitochondrial_cytb
genotypes_resistance %>%
  filter(Gene == "mitochondrial_cytb") %>%
  dplyr::count(AA_change, Continent,Sampling_Date, Allele_type) %>%
  pivot_wider(names_from = Allele_type, values_from = n, values_fill = list(n = 0)) %>%
  mutate(Number_all = Mutant + Origin) %>%
  mutate(prop_mutant = Mutant/Number_all) %>%
  group_by(AA_change) %>%
  dplyr::mutate(Somme = sum(Mutant)) %>%
  filter(Somme > 0) %>%
  ggplot(aes(x=Sampling_Date, y=Continent, size = prop_mutant, fill = prop_mutant)) +
  geom_point(shape=21, alpha =0.7) + theme_bw() +
  facet_wrap(.~AA_change) + scale_fill_gradient(low="white", high = "#16697a") +
  labs(title = str_wrap(paste("Mutant proportion per known mutation",
                              "in the mitochondrial gene cytb"), width = 35),
       subtitle = str_wrap(paste("This gene can mutate and cause resistance ",
                                 "to QoI, or Quinone outside inhibitors."), width = 65),
       fill = "Proportion of mutants",
       size = "Proportion of mutants",
       x = "Years", y = "")


# SDH genes
genotypes_resistance %>%
  filter(grepl("SDH", Gene)) %>%
  unite(Label, Gene, AA_change, remove = T) %>%
  dplyr::count(Label, Continent, Sampling_Date, Allele_type) %>%
  pivot_wider(names_from = Allele_type, values_from = n, values_fill = list(n = 0)) %>%
  mutate(Number_all = Mutant + Origin) %>%
  mutate(prop_mutant = Mutant/Number_all) %>%
  group_by(Label) %>%
  dplyr::mutate(Somme = sum(Mutant)) %>%
  filter(Somme > 0) %>%
  ggplot(aes(x=Sampling_Date, y=Continent, size = prop_mutant, fill = prop_mutant)) +
  geom_point(shape=21, alpha =0.7) + theme_bw() +
  facet_wrap(.~Label) + scale_fill_gradient(low="white", high = "#16697a")+
  labs(title = str_wrap(paste("Mutant proportion per known mutation",
                              "in one of the SDH genes"), width = 35),
       subtitle = str_wrap(paste("This gene can mutate and cause resistance to SDHI fungicides."), width = 65),
       fill = "Proportion of mutants",
       size = "Proportion of mutants",
       x = "Years", y = "")
```



<br><br>


# Virulence to new varieties
***

## Thierry Marcel's virulence GWAS
I use the alleles identified in the GWAS of Thierry Marcel and track them in the world-wide populations.
```{r French GWAS Virulence}
virulence_table = read_tsv(paste0(virulence_dir, "French_GWAS_virulence_alleles.tab")) %>%
  separate(VIRULENCE_ALLELES, into = c("VIRULENCE_ALLELE1", "VIRULENCE_ALLELE2"), sep = ",")

french_GWAS_variants = read_tsv(paste0(virulence_dir, "Positions_GWAS.short.vcf"),
                                na = ".")  %>%
  pivot_longer(-c(`#CHROM`, POS, REF, ALT), names_to = "ID_file", values_to = "Allele_nb") %>%
  separate(ALT, into = c("ALT1", "ALT2"), sep = ",") %>%
  mutate(Allele_ID = ifelse(Allele_nb == 0, REF, ifelse(Allele_nb == 1, ALT1, ALT2))) %>%
  left_join(., virulence_table) %>%
  mutate(virulence = ifelse(Allele_ID == NON_VIRULENT_ALLELES, "Non_virulent",
                            ifelse(Allele_ID == VIRULENCE_ALLELE1 | Allele_ID == VIRULENCE_ALLELE2,
                                   "Virulent", NA)))

french_GWAS_variants = inner_join(french_GWAS_variants, Zt_meta)


french_GWAS_variants %>%
  tidyr::unite(position, `#CHROM`, POS) %>%
  dplyr::count(position, Continent,Sampling_Date, virulence) %>%
  pivot_wider(names_from = virulence, values_from = n, values_fill = list(n = 0)) %>%
  mutate(Number_all = Virulent + Non_virulent) %>%
  mutate(prop_virulent = Virulent/Number_all) %>%
  ggplot(aes(x=Sampling_Date, y=Continent, size = prop_virulent, fill = prop_virulent)) +
  geom_point(shape=21, alpha =0.7) + theme_bw() +
  scale_fill_gradient(low="white", high = "#16697a")+
  labs(title = str_wrap(paste("Virulence alleles",
                              ""), width = 35),
       subtitle = str_wrap(paste("These alleles were identified in a GWAS study",
                                 " to increase either virulence or aggressiveness",
                                 " on some wheat varieties."), width = 65),
       fill = "Proportion of virulent",
       size = "Proportion of virulent",
       x = "Years", y = "") +
  facet_wrap(vars(position))

```
Welll... Whatever...

## Avr_Stb6
```{bash, eval = F}
#On the cluster: run blast
while read sample ; do sh Detect_gene_blast.sh ./Directories_new.sh /data2/alice/WW_project/7_Virulence/Avr_Stb6.fasta ${sample} Illumina /data2/alice/WW_project/7_Virulence/${sample}_Avr_Stb6.blast.tab; done < list_Third_batch_Dec_2018.txt

#On the cluster: Gather blast results
cat /data2/alice/WW_project/7_Virulence/*Avr_Stb6.blast.tab > /data2/alice/WW_project/7_Virulence/Overall_results_blast_Avr_Stb6.txt

#On the cluster: Gather blast results
while read p ;
 do
sample=$(echo $p | cut -f 1 -d " " ) ;
 chr=$(echo $p | cut -f 4 -d " ") ;
 start=$(echo $p | cut -f 11 -d " ") ; end=$(echo $p | cut -f 12 -d " ") ;
 echo $sample, $chr, $start, $end ;
 order_start=$(echo -e "$start\n$end"  | sort -n | head -n1);
 order_end=$(echo -e "$start\n$end"  | sort -nr | head -n1);
 echo -e "${chr}\t${order_start}\t${order_end}" > temp.bed ;
 ~/Software/bedtools getfasta \
   -fi /data2/alice/WW_project/4_TE_RIP/1_Blast_from_denovo_assemblies/0_Spades/${sample}.fasta\
   -bed temp.bed | \
   sed "s/>/>${sample}:/" >> \
 /data2/alice/WW_project/7_Virulence/Overall_results_blast_Avr_Stb6.fasta ;
done < /data2/alice/WW_project/7_Virulence/Overall_results_blast_Avr_Stb6.txt

#And on the website, because laziness
mafft --thread 8 --threadtb 5 --threadit 0 --reorder --adjustdirection --auto input > output
```

Here is the tree for the Avr_Stb6 gene. For reference, the allele carried by ST99CH_1A4 gives an avirulent phenotype, while the allele from ST99CH_1A5 is the virulent allele (from [Zhong et al. 2017](https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.14434)).

```{r AvrStb6 tree plots, eval = F}
#grep ">" ../7_Virulence/Overall_results_blast_Avr_Stb6.fasta | sed 's/>//' > ../7_Virulence/Overall_results_blast_Avr_Stb6.list

temp2 = read_tsv(paste0(virulence_dir, "Overall_results_blast_Avr_Stb6.list"), col_names = "Fasta_header") %>%
  separate(Fasta_header, into = c("ID_file", "chr", "coords"), sep=":", remove = F) %>%
  mutate(contig2 = stringr::str_replace_all(Fasta_header, "[:.]", "_")) %>%
  left_join(., Zt_meta)

#Read tree in
#details from the mafft website about the tree
#Size = 640 sequences × 319 sites
#Method = Neighbor-Joining
#Model = Jukes-Cantor
#Bootstrap resampling = 0
#Alignment id = .200930173318269H5fPcXaMbnDOkoF7E0Y3ilsfnormal


tree_file = "Tree_Avr_Stb6.nwk"
tree = as_tibble(treeio::read.tree(paste0(virulence_dir, tree_file))) %>%
  mutate(label_copy = label) %>%
  separate(label_copy, into = c("nb", "contig"), extra = "merge", remove =F) %>%
  dplyr::mutate(nb = as.integer(nb)) %>%
  dplyr::mutate(label_OG = label) %>%
  dplyr::mutate(contig2 = stringr::str_replace(contig, "R_", "")) %>%
  left_join(., temp2, by = c("contig2")) %>%
  dplyr::mutate(label_temp = ifelse(is.na(ID_file), ifelse(is.na(contig), label, contig), ID_file)) %>%
  unite(col = "label_new", nb, label_temp, sep = "_")

tree2 = as_tibble(treeio::read.tree(paste0(virulence_dir, tree_file)))

temp = tree %>%
  mutate(original_vir = ifelse(ID_file == "ST99CH_1A5", "ST99CH_1A5",
  ifelse(ID_file == "ST99CH_1E4", "ST99CH_1E4", NA))) %>%
  dplyr::select(label, node, label_new, Continent, Sampling_Date, Collection, original_vir) %>%
  filter(!is.na(label))

#Find the outgroup!
root_node = tree2 %>%
  filter(grepl("Zp13", label)) %>%
  dplyr::select(node) %>%
  pull()


rooted.tree <- ape::root(as.treedata(tree2), root_node)
p <- ggtree(rooted.tree, layout = "rectangular") %<+% (temp %>% dplyr::select (-label))

p2 <- p + geom_tippoint(aes(color = Continent), alpha = 0.7) +
  theme(legend.position = "right") +
  Color_Continent

#p2

p3 <- p + geom_tippoint(aes(color = original_vir)) +
  theme(legend.position = "right")  +
  scale_color_manual(values = c(mycolorsCorrel[1], mycolorsCorrel[20]))
cowplot::plot_grid(p2, p3)


p4 <- p + geom_tippoint(aes(color = Sampling_Date), alpha = 0.7) +
  theme(legend.position = "right")
cowplot::plot_grid(p4, p3)
```
