---
title: "Per windows statistics: variant calling Z.tritici"
output: 
   html_document:
    code_folding: hide
---


```{r setup, warning=FALSE, message = FALSE}

library(knitr)
library(reticulate)

#Data wrangling and data viz
library(tidyverse)
library(purrr)
library(RColorBrewer)
library(plotly)
library(cowplot)
library(GGally)
library(corrplot)
library(ggstance)
library('pophelper')
library(ggbiplot)
library(igraph)
library(ggraph)
library(ggrepel)
library(ggtext)
library(scatterpie)
library(pheatmap)
library(ggridges)


library(ggtree)
library(tidytree)
library(multcomp)
library(lsmeans)

#Statistics
library(car)
library(corrr)
library(lsmeans)
library(multcomp)

#Variables
world <- map_data("world")
project_dir="~/Documents/Postdoc_Bruce/Projects/WW_project/"
lists_dir = "~/Documents/Postdoc_Bruce/Projects/WW_project/WW_PopGen/Keep_lists_samples/"

#Data directories
data_dir=paste0(project_dir, "0_Data/")
metadata_dir=paste0(project_dir, "Metadata/")
fig_dir = "~/Documents/Postdoc_Bruce/Manuscripts/Feurtey_WW_Zt/Draft_figures/"

#Analysis directories
#-___________________
VAR_dir = paste0(project_dir, "1_Variant_calling/")
  depth_per_window_dir = paste0(VAR_dir, "1_Depth_per_window/")
  depth_per_gene_dir = paste0(VAR_dir, "2_Depth_per_gene/")
  vcf_dir = paste0(VAR_dir, "4_Joint_calling/")
  mito_SV = paste0(VAR_dir, "6_Mito_SV/")
  chipseq_dir = paste0(VAR_dir, "7_ChipSeq_peaks/")
PopStr_dir = paste0(project_dir, "2_Population_structure/")
  nuc_PS_dir=paste0(PopStr_dir, "0_Nuclear_genome/")
  mito_PS_dir = paste0(PopStr_dir, "1_Mitochondrial_genome/")
Sumstats_dir = paste0(project_dir, "3_Sumstats_demography/")
TE_RIP_dir=paste0(project_dir, "4_TE_RIP/")
   RIP_DIR=paste0(TE_RIP_dir, "0_RIP_estimation/")
   DIM2_DIR=paste0(TE_RIP_dir, "1_Blast_from_denovo_assemblies/")
GEA_dir=paste0(project_dir, "5_GEA/")
fung_dir=paste0(project_dir, "6_Fungicide_resistance/")
virulence_dir = paste0(project_dir, "7_Virulence/")
sel_dir = paste0(project_dir, "8_Selection/")
  gene_list_dir = paste0(sel_dir, "0_Lists_unique_copy/")


#Files
vcf_name="Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.biall_SNP.max-m-1.maf-0.05.thin-1000bp"
vcf_name_nomaf="Ztritici_global_March2021.filtered-clean.AB_filtered.SNP.max-m-0.8.thin-1000bp"
vcf_name_mito = "Ztritici_global_March2021.genotyped.mt.filtered.clean.AB_filtered.variants.good_samples.max-m-80"
Zt_list = paste0(lists_dir, "Ztritici_global_March2021.genotyped.good_samples.args")
effectors_annotation_file = "~/Documents/Postdoc_Eva/Manuscripts/Accepted/Alice_Cecile_Comparative_genomics/Data_for_publication/Annotations_2018_genomes_for_publication.tab"
eggnog_annotation = paste0(data_dir, "Zymoseptoria_tritici.MG2.Grandaubert2015.eggnog")
gff_file = paste0(data_dir, "Zymoseptoria_tritici.MG2.Grandaubert2015.no_CDS.gff3")
ref_fasta_file = paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa")
metadata_name = "Main_table_from_SQL_Feb_2020"
gene_annotation = read_tsv(paste0(data_dir, "Badet_GLOBAL_PANGENOME_TABLE.txt"))
complete_mito = read_tsv(paste0(data_dir, "Complete_mitochondria_from_blast.txt"), col_names = c("ID_file", "Contig"))

Sys.setenv(PROJECTDIR=project_dir)
Sys.setenv(VARDIR=VAR_dir)
Sys.setenv(VCFDIR=vcf_dir)
Sys.setenv(POPSTR=PopStr_dir)
Sys.setenv(MITOPOPSTR=mito_PS_dir)
Sys.setenv(TERIP=TE_RIP_dir)

Sys.setenv(SUMST=Sumstats_dir)
Sys.setenv(GEADIR=GEA_dir)

Sys.setenv(ZTLIST=Zt_list)
Sys.setenv(GFFFILE = gff_file)
Sys.setenv(REFFILE = ref_fasta_file)
Sys.setenv(VCFNAME=vcf_name)
Sys.setenv(VCFNAME_NOMAF=vcf_name_nomaf)
Sys.setenv(VCFNAME_MITO=vcf_name_mito)

#knitr::opts_chunk$set(echo = F)
knitr::opts_chunk$set(message = F)
knitr::opts_chunk$set(warning = F)
knitr::opts_chunk$set(results = T)


# Metadata and sample lists
##Filtered_samples
filtered_samples = bind_rows(
  read_tsv(paste0(metadata_dir, "Sample_removed_based_on_IBS.args"), col_names = "ID_file") %>%
  mutate(Filter = "IBS"),
read_tsv(paste0(metadata_dir, "Sample_with_too_much_NA.args"), col_names = "ID_file") %>%
  mutate(Filter = "High_NA"),
read_tsv(paste0(metadata_dir, "Samples_to_filter_out.args"), col_names = "ID_file") %>%
  mutate(Filter = "Mutants_etc"))

##Samples in vcf
genotyped_samples = read_tsv(Zt_list, col_names = "ID_file")

## Metadata of genotyped samples 
temp = read_tsv(paste0(metadata_dir, metadata_name, "_Description.tab"), col_names = F) %>% pull()

Zt_meta = read_delim(paste0(metadata_dir, metadata_name, "_with_collection.tab"), 
                 col_names = temp,
                 na = "\\N", guess_max = 2000) %>%
  unite(Coordinates, Latitude, Longitude, sep = ";", remove = F) %>%
  inner_join(., genotyped_samples)  %>%
  mutate(Country = ifelse(Country == "USA", paste(Country, Region, sep = "_"), Country)) %>%
  mutate(Country = ifelse(Country == "Australia", paste(Country, Region, sep = "_"), Country)) %>%
  mutate(Country = ifelse(Country == "NZ", "New Zealand", Country)) %>%
  mutate(Country = ifelse(Country == "CH", "Switzerland", Country)) %>%
  mutate(Latitude2 = round(Latitude, 2), Longitude2 = round(Longitude, 2)) %>%
  dplyr::select(ID_file, Continent, Country, Latitude, Longitude, Latitude2, Longitude2,
                Sampling_Date, Collection)

temp = read_tsv(paste0(PopStr_dir, vcf_name, ".high_anc_coef_snmf.tsv")) %>%
  dplyr::select(ID_file = Sample, Cluster)

Zt_meta = full_join(Zt_meta, temp)

#genotyped_samples %>%
#  filter(!(ID_file %in% filtered_samples$ID_file)) %>%
#    write_tsv(Zt_list, col_names = F)



#Define colors
## For continents
#myColors <- c("#04078B", "#a10208", "#FFBA08", "#CC0044", "#5C9D06", "#129EBA","#305D1B")
myColors <- c("#DA4167", "grey", "#ffba0a", "#A20106", "#3F6E0C", "#129eba", "#8fa253" )
names(myColors) = levels(factor(Zt_meta$Continent))
Color_Continent = ggplot2::scale_colour_manual(name = "Continent", values = myColors)
Fill_Continent = ggplot2::scale_fill_manual(name = "Continent", values = myColors)

myColors2 <- c("#129eba", "#3F6E0C", "#DA4167", "#ffba0a", "#129eba", "#129eba", 
               "#8fa253", "#A20106", "#A20106", "#3F6E0C", "#8fa253")
names(myColors2) = levels(factor(Zt_meta$Cluster))
Color_Cluster = ggplot2::scale_colour_manual(name = "Cluster", values = myColors2)
Fill_Cluster = ggplot2::scale_fill_manual(name = "Cluster", values = myColors2)

## For clustering
K_colors = c("#f9c74f", "#f9844a", "#90be6d", "#f5cac3", 
"#83c5be", "#f28482", "#577590", "#e5e5e5", "#a09abc",  "#52796f",
"#219ebc", "#003049", "#82c0cc", "#283618", "white")

## For correlations
mycolorsCorrel<- colorRampPalette(c("#0f8b8d", "white", "#a8201a"))(20)

#Random gradients
greens=c("#1B512D", "#669046", "#8CB053", "#B1CF5F", "#514F59")
blues=c("#08386E", "#1C89C9", "#28A7C0", "#B0DFE8", "grey")
```

```{bash perwin stats, eval =F}
#Run on the cluster

#Create bed file with 10kb windows
#(including the last window which can be smaller)
while read chr length temp temp2 temp3; do 
  start=0; 
  while [ "$start" -le "$length" ] ; do 
    if [ "$(($start + 10000))" -le  "$length" ] ; 
    then 
      echo -e "${chr}\t${start}\t$(($start + 10000))" ; 
    else  echo -e "${chr}\t${start}\t$length" ;  
    fi ; 
    start=`expr $start + 10000` ; 
  done ; 
done < /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa.fai > /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed


#GC etc from bedtools nuc
#columns are "CHROM", "Start", "End", "ATpercent", "GCpercent", "NbA", "NbC", "NbG, "NbT", "NbN", "NbOther", "Length"

~/Software/bedtools nuc \
    -fi /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa \
    -bed /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed \
    > /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.nuc_GC.tab

#Mappability
genmap index -F /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa -I /data2/alice/WW_project/0_Data/Zymoseptoria_tritici_genmap_index

genmap map -K 30 -E 2 -I  /data2/alice/WW_project/0_Data/Zymoseptoria_tritici_genmap_index -O /data2/alice/WW_project/0_Data/Zymoseptoria_tritici_genmap_mappability -t -bg

#RIP per window 
#(my script makes a summary for all seq in a multifasta, so I tricked it my giving it a single window at a time)
 #columns are "CHROM", "Start", "End", "GC", "Product index", "Substrate index", "Composite"
 while read chr start end ; 
    do 
    echo -e "${chr}\t${start}\t${end}" > temp.bed ; 
    ~/Software/bedtools getfasta -fi /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.fa \
       -bed temp.bed -fo temp.fa ; 
    python GC_RIP_per_read_fastq.py --input_format fasta --out temp temp.fa ; 
    values=$(cat temp.txt | ~/Software/datamash-1.3/datamash transpose | grep "Median" | cut -f 2,3,4,5) ; 
    echo -e "${chr}\t${start}\t${end}\t$values" \
    >> /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.RIP.tsv ; 
done < /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed


#Count variants
#columns are CHROM, Start(IN 10KB UNITS), Variant_number
zcat /data2/alice/WW_project/1_Variant_calling/4_Joint_calling/Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.max-m-80.vcf.gz  | \
   grep -v "#"  | awk 'BEGIN {FS="\t"; OFS="\t"} {print $1,$2,int($2/10000)}' | \
   ~/Software/bedtools groupby -g 1,3 -o count -c 2 > \
   /data2/alice/WW_project/1_Variant_calling/4_Joint_calling/Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.max-m-80.10kb_windows.SNP_counts.tab
   
   
#Gene coverage
#columns are "CHROM", "Start", "End", "Nb_overlapping_genes", "Coverage_bp_gene", "Window_length", "Coverage_frac_gene"
~/Software/bedtools coverage \
   -a /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed \
   -b /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.Grandaubert2015.mRNA.gff3 \
   > /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.gene_coverage.tab
 
 #Reference TE coverage
#columns are "CHROM", "Start", "End", "Nb_overlapping_TEs", "Coverage_bp_TE", "Window_length", "Coverage_frac_TE"
 ~/Software/bedtools coverage \
    -a /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed \
    -b /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.Badet_Oggenfuss_2019.TE.gtf \
    > /data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.TE_coverage.tab
```


```{bash Importation from cluster, eval = F}
# Uploading commands from the cluster to my computer.

#Per windows: coordinates, RIP, GC, SNP counts
rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed \
 ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed
 
rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.RIP.tsv \
 ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.RIP.tsv 
 
rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.nuc_GC.tab \
 ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.nuc_GC.tab
 
rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/1_Variant_calling/4_Joint_calling/Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.max-m-80.10kb_windows.SNP_counts.tab  \
 ~/Documents/Postdoc_Bruce/Projects/WW_project/1_Variant_calling/4_Joint_calling/
 
rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.gene_coverage.tab \
~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/

rsync -avP   alice@130.125.25.244:/data2/alice/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.TE_coverage.tab \
 ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/
 
rsync -avP alice@130.125.25.244:/data2/alice/WW_project/1_Variant_calling/7_ChipSeq_peaks ../1_Variant_calling/
```


```{bash merge methyl, eval = F}

#This is done for each histone mark
#The data is from Klaas paper's on centromeres
#The peaks were called by macs2 (option --nomodel) after trimming and mapping with trimmomatic and bowtie + filtering of reads mapping with a quality sup to 30

grep -v "#" ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep1_SRR2060839_peaks.xls  | cut -f 1,2,3,10 | \
   grep "peak" > ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep1_SRR2060839_peaks.bed
grep -v "#" ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep2_SRR2060840_peaks.xls  | cut -f 1,2,3,10 | \
   grep "peak" > ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep2_SRR2060840_peaks.bed

~/Documents/Software/bedtools2/bin/bedtools intersect \
   -a ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep1_SRR2060839_peaks.bed \
   -b ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_Rep2_SRR2060840_peaks.bed \
   > ../1_Variant_calling/7_ChipSeq_peaks/H3K9me3_intersect_only_overlap.bed 



#Get length of overlap between genes and windows for RNAseq
~/Documents/Software/bedtools2/bin/bedtools intersect -wo \
  -a ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed \
  -b ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.Grandaubert2015.gff3  \
  > ~/Documents/Postdoc_Bruce/Projects/WW_project/0_Data/Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.intersect_genes.txt
```

```{r}

#Getting mappability per window
mappability = read_tsv(paste0(data_dir, "Zymoseptoria_tritici_genmap_mappability.bedgraph"), 
         col_names = c("CHR", "start", "end", "mappab"), col_types = c("c", "d", "d", "d")) %>%
  dplyr::mutate(Length = (end - start), window = trunc(((start + end)/2)/10000)) %>%
  group_by(CHR, window) %>%
  dplyr::summarize(Mappability = weighted.mean(x = mappab, w = Length)) %>% 
  mutate(Start = window*10000) %>%
  ungroup() %>%
  dplyr::select(-window)


#Getting the methylation peaks
chipseq = bind_rows(read_tsv(paste0(chipseq_dir, "H3K27me3_intersect_only_overlap.bed"), 
         col_names = c("CHR", "start", "end", "peak"), col_types = c("c", "d", "d", "c")),
    read_tsv(paste0(chipseq_dir, "H3K4me2_intersect_only_overlap.bed"), 
         col_names = c("CHR", "start", "end", "peak"), col_types = c("c", "d", "d", "c"))) %>%
    bind_rows(read_tsv(paste0(chipseq_dir, "H3K9me2_intersect_only_overlap.bed"), 
         col_names = c("CHR", "start", "end", "peak"), col_types = c("c", "d", "d", "c"))) %>%
      separate(col = peak, into = c("Mark"), sep ="_", remove = FALSE, extra = "drop") %>%
  dplyr::mutate(Length = (end - start), window = trunc(((start + end)/2)/10000)) %>%
  group_by(CHR, window, Mark) %>%
  dplyr::summarize(Coverage = sum(Length)/10000) %>% 
  mutate(Start = window*10000) %>%
  ungroup() %>%
  dplyr::select(-window) %>%
  pivot_wider(names_from = Mark, values_from = Coverage, values_fill = 0)



#RNAseq
TPM = read_tsv(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.intersect_genes.txt"), 
         col_names = c("CHR", "Start", "End", "X1", "X2", "X3", "Gene_start", "Gene_end",
                       "X4", "X5", "X6", "Annot", "Overlap")) %>%
  dplyr::select(CHR, Start, End, Annot, Overlap) %>%
  separate(col = Annot, into = c("X1", "X2", "X3", "X4","X5", "Protein_ID"), sep = "[=;]")  %>%
  inner_join(readxl::read_excel(paste0("/Users/afeurtey/Documents/Postdoc_Eva/Manuscripts/",
                                       "Accepted/Alice_Cecile_Comparative_genomics/",
                                       "Data_for_publication/Expression_data_Ztritici_inplanta.xlsx"), 
                                skip = 5, sheet = "IPO323_TPM")) %>%
  group_by(CHR, Start, End, Protein_ID) %>%
  mutate(Average_bio = mean(c(A_1, A_2, B_1, B_2), na.rm = T),
         Average_necro = mean(c(C_1, C_2, D_1, D_2)), na.rm = T,
         CHR = as.character(CHR)) %>%
  dplyr::select(CHR, Start, End, Protein_ID, Overlap, Average_bio, Average_necro) 

TPM_per_window = TPM %>%
  filter(Protein_ID != "Zt09_3_00018") %>%
  group_by(CHR, Start, End) %>%
  dplyr::summarize(TPM_bio = weighted.mean(x = Average_bio, w = Overlap),
                   TPM_necro = weighted.mean(x = Average_necro, w = Overlap)) 
``` 


# PCA and mappability 

I want to explore the variation of the different genomic characteristics per windows. I also want to check the effect of mappability on the patterns we are seeing, as it is expected to be the main bias here. 
```{r initial PCA}
#Getting all the data together
data = read_delim(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.bed"), 
                  col_names = c("CHR", "Start", "End")) %>%
full_join(read_delim(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.nuc_GC.tab"), skip = 1,
         col_names = c("CHR", "Start", "End", "ATpercent", "GCpercent", "NbA", "NbC", "NbG", "NbT", 
                       "NbN", "NbOther", "Window_length"))) %>%
  dplyr::select(CHR, Start, End, GCpercent, Window_length) %>%
full_join(read_delim(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.RIP.tsv"), 
         col_names = c("CHR", "Start", "End", "GC", "Product_index", "Substrate_index", "Composite_index"),
         na = c("", "nan"), delim = "\t"))  %>%
full_join(read_tsv(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.gene_coverage.tab"), 
         col_names = c("CHR", "Start", "End", "Nb_overlapping_genes", "Coverage_bp_gene", "Window_length", "Coverage_frac_gene"))) %>%
full_join(read_tsv(paste0(data_dir, "Zymoseptoria_tritici.MG2.dna.toplevel.mt+.10kb_windows.TE_coverage.tab"), 
         col_names = c("CHR", "Start", "End", "Nb_overlapping_TEs", "Coverage_bp_TE", "Window_length", "Coverage_frac_TE"))) %>%
full_join(read_delim(paste0(TE_RIP_dir, "Ztritici_global_March2021.good_samples.10kb_windows.TIP_counts.tab"),
                     col_types = c("c","d","c","d","d"))) %>%
full_join(read_tsv(paste0(vcf_dir, 
                 "Ztritici_global_March2021.genotyped.ALL.filtered.clean.AB_filtered.variants.good_samples.max-m-80.10kb_windows.SNP_counts.tab"),
          col_names = c("CHR", "Start", "Short_variant_number"), col_types = c("c","d","d")) %>% 
            mutate(Start = Start*10000)) %>%
  full_join(mappability) %>%
  full_join(chipseq) %>%
  full_join(TPM_per_window)


#PCA computation
data_PCA = data %>% 
  unite(CHR, Start, col = Window, sep = ":") %>%
  filter(Window_length > 7500) %>%
  #filter(TIP_count < 150) %>%
  dplyr::select(Window, GC, Composite_index, Coverage_frac_gene, Coverage_frac_TE, 
                TIP_count, Short_variant_number, Mappability, 
                H3K27me3, H3K9me2, H3K4me2, TPM_bio, TPM_necro) 
data_PCA[is.na(data_PCA)] <- 0

temp = as.matrix(data_PCA[,c(2:(ncol(data_PCA)-1))])[, which(apply(as.matrix(data_PCA[,c(2:(ncol(data_PCA)-1))]), 2, var) != 0)]

per_win.pca = prcomp(temp, center = TRUE,scale. = TRUE)

temp = as.tibble(cbind(data_PCA, as.data.frame(per_win.pca$x)))


#PCA visualization: coloring dots with mappability
arrows = as_tibble(as.data.frame(per_win.pca$rotation)) %>%
  mutate(label = rownames(per_win.pca$rotation))


for_plot = temp %>%
  separate(col = Window, into = c("CHR"), remove = F, extra = "drop", sep = ":") %>%
  mutate(CHR_type = ifelse(as.numeric(CHR) < 14, "Core", "Accessory"))

ggplot() + 
    geom_point(data = temp, aes(x = PC1, y = PC2, col = Mappability), alpha = .6, shape = 1) + 
    theme_bw() + 
    geom_segment(data = arrows, aes(x = 0, y = 0, xend = PC1*6, yend = PC2*6 ),
                 arrow = arrow(length = unit(0.03, "npc"))) +
    geom_text(data = arrows, aes(x = PC1*6, y = PC2*6, label = label)) + 
    coord_cartesian(xlim = c(-4, 6)) +
  scale_color_gradient(high = "#ADA7A9", low = "#ffa62b") +
  theme(legend.position = "bottom")

```
Not surprisingly the mappability is strongly linked to the first PC, as windows with high TE converage and high composite index tend to have lower mappability than gene rich loci. I'll look more into this here, by placing a threshold of mappability to remove low mappability windows. I visualize the distribution of the variables before and after filtering on the mappability threshold.
```{r filter mappability}

#Threshold setting, visualizing and data filtering
threshold = 0.85
data_PCA %>%
  ggplot(aes(Mappability)) + 
  geom_density(fill = "#EDE7E3", col ="#ADA7A9") + 
  geom_vline(xintercept = threshold, col = "#82c0cc") + 
  theme_bw()
data_PCA2 = data_PCA %>%
  filter(Mappability > threshold) 


#First, density plots of pre and post filtering
pre_post_filter = bind_rows(data_PCA %>%
    pivot_longer(cols = -Window, names_to = "Measure", values_to = "Estimate") %>% 
    mutate(Windows = "All"),
          data_PCA2 %>%
      pivot_longer(cols = -Window, names_to = "Measure", values_to = "Estimate")  %>% 
      mutate(Windows = "Only mappable")) %>%
  filter(Measure != "Mappability") %>%
  group_by(Measure) %>%
  dplyr::mutate(Med = median(Estimate)) %>%
  mutate(Variable_category = case_when(str_detect(Measure, "H3") ~ "Epigenomics", 
                                       Measure == "Coverage_frac_TE" ~ "TE related",
                                       str_detect(Measure, "TPM") ~ "Gene related",
                                       Measure == "Coverage_frac_gene" ~ "Gene related",
                                       Measure %in% c("GC", "Composite_index") ~ "TE related",
                                       TRUE ~ "Variant counts")) 

ggplot(pre_post_filter, aes(x = Estimate, col = Windows, fill = Windows)) +
  geom_density(alpha = .6) + 
  theme_bw() + 
  theme(panel.grid = element_blank()) + 
  facet_wrap(vars(Measure), scales = "free")  +
  scale_color_manual(values = c("grey", "#2FC6B7"))+
  scale_fill_manual(values = c("grey", "#2FC6B7")) 


#Violing and boxplot of the short variant and TIP counts
pre_post_filter %>%
  filter(Variable_category == "Variant counts") %>%
  ggplot(aes(y = Estimate, x = Windows, 
             fill = Windows, col = Windows)) +
    geom_violin(alpha = .4) + 
    geom_boxplot(alpha = .4, width = .2, outlier.shape = NA) + 
    theme_bw() +
    theme(panel.grid.major.x = element_blank()) +
    labs(y = "Variant count per 10kb window", y = "") +
  scale_color_manual(values = c("grey", "#2FC6B7"))+
  scale_fill_manual(values = c("grey", "#2FC6B7")) + 
  facet_wrap(vars(Measure), scales = "free", ncol = 3)  

ggsave(paste0(fig_dir, "FigS2_variants_filtered_on_mappability.pdf"), width = 12, height = 12, units = "cm")

```

Some variables are particularly related to mappability. Here, I will compare the values between the high and low mappability windows.
```{r comparison high and low mappability}
#Boxplots for a subset of the variables

temp = data %>% 
  filter(!is.na(Mappability)) %>%
  mutate(Mappability = ifelse(Mappability < threshold, "Low", "High"))
  

#TE coverage
p1 = ggplot(temp, aes(x = Mappability, y = Coverage_frac_TE, col = Mappability, fill = Mappability)) +
     geom_boxplot() +
     theme_bw()+
     scale_color_manual(values = c("#ADA7A9", "#ffa62b")) + 
     scale_fill_manual(values = c("#EDE7E3", "#FFD399"))

#H3K27me3 mark
 p2 = ggplot(temp, aes(x = Mappability, y = H3K27me3, col = Mappability, fill = Mappability)) +
      geom_boxplot(alpha = .4) +
      theme_bw() +
      scale_color_manual(values = c("#ADA7A9", "#ffa62b")) + 
      scale_fill_manual(values = c("#EDE7E3", "#FFD399"))

#Gene coverage
p3 = ggplot(temp, aes(x = Mappability, y = Coverage_frac_gene, col = Mappability, fill = Mappability)) +
     geom_boxplot() +
     theme_bw() +
     scale_color_manual(values = c("#ADA7A9", "#ffa62b")) + 
     scale_fill_manual(values = c("#EDE7E3", "#FFD399"))

#Composite index
p4 = data %>% 
  filter(!is.na(Mappability)) %>%
  mutate(Mappability = ifelse(Mappability < threshold, "Low", "High")) %>%
  ggplot(aes(x = Mappability, y = Composite_index, col = Mappability, fill = Mappability)) +
    geom_boxplot() +
    theme_bw() +
  scale_color_manual(values = c("#ADA7A9", "#ffa62b")) + 
  scale_fill_manual(values = c("#EDE7E3", "#FFD399"))



 cowplot::plot_grid(p1 + theme(legend.position = "none", axis.title.x = element_blank()), 
                    p2 + theme(legend.position = "none", axis.title.x = element_blank()),
                    p3 + theme(legend.position = "none"), p4 + theme(legend.position = "none"),
                    rel_heights = c(1, 1, 1.25, 1.25))

ggsave(paste0(fig_dir, "FigS2_boxplots_high_low_mappability.pdf"), width = 12, height = 12, units = "cm")

```


# PCA after filtering on mappability

Here are the distributions of the different variables in the filtered data. 
```{r filtered data}
 

pre_post_filter$Variable_category = factor(pre_post_filter$Variable_category, 
                                           levels = c("Gene related", "TE related", "Epigenomics"))
pre_post_filter %>%
  filter(Windows == "Only mappable") %>%
  filter(Variable_category != "Variant counts") %>%
  ggplot(aes(x = Estimate)) +
    geom_density(alpha = .4, fill = "grey", color = "grey") + 
    theme_bw() +
    theme(panel.grid.minor.x = element_blank(), panel.grid.minor.y = element_blank()) +
    labs(x = "Coverage proportion per window", y = "Type of feature") +
  facet_wrap(vars(Variable_category, Measure), scales = "free", ncol = 3) 

ggsave(paste0(fig_dir, "Fig1_Distribution.pdf"), width = 12, height = 12, units = "cm")

```

Now, I want to redo the PCA without taking into account windows with low mappability. I also color the dots this time with the type of chromosomes, accessory chromosomes vs core chromosomes.
```{r filtered PCA}
temp = data_PCA2 %>% 
  dplyr::select(-Mappability) 

temp = as.matrix(temp[,c(2:ncol(temp))])[, which(apply(as.matrix(temp[,c(2:ncol(temp))]), 2, var) != 0)]

per_win.pca = prcomp(temp, center = TRUE,scale. = TRUE)

#Making custom PCA plot
arrows = as_tibble(as.data.frame(per_win.pca$rotation)) %>%
  mutate(label = rownames(per_win.pca$rotation))

temp = as.tibble(cbind(data_PCA2, as.data.frame(per_win.pca$x)))

for_plot = temp %>%
  separate(col = Window, into = c("CHR"), remove = F, extra = "drop", sep = ":") %>%
  mutate(CHR_type = ifelse(as.numeric(CHR) < 14, "Core", "Accessory"))
eigs <- per_win.pca$sdev^2

ggplot() + 
    geom_point(data = for_plot, aes(x = PC1, y = PC2, col = CHR_type), alpha = .6, shape = 1) + 
    theme_bw() + 
    geom_segment(data = arrows, aes(x = 0, y = 0, xend = PC1*10, yend = PC2*10 ),
                 arrow = arrow(length = unit(0.03, "npc"))) +
    geom_text(data = arrows, aes(x = PC1*10, y = PC2*10, label = label)) + 
  coord_cartesian(xlim = c(-6, 10), ylim = c(-8, 7)) +
  scale_color_manual(values = c("#ffa62b", "#2FC6B7")) +
  labs(title = "PCA per 10kb windows", 
       subtitle = "Two first PCs",
       x = paste0("PC1 (", round(100*eigs[1]/sum(eigs), 1), "% explained var.)"),
       y = paste0("PC2 (", round(100*eigs[2]/sum(eigs), 1), "% explained var.)"))

ggsave(paste0(fig_dir, "Fig1_PCA_per_10kb.pdf"), width = 20, height = 12, units = "cm")


ggplot() + 
    geom_point(data = for_plot, aes(x = PC3, y = PC4, col = CHR_type), alpha = .6, shape = 1) + 
    theme_bw() + 
    geom_segment(data = arrows, aes(x = 0, y = 0, xend = PC3*10, yend = PC4*10 ),
                 arrow = arrow(length = unit(0.03, "npc"))) +
    geom_text(data = arrows, aes(x = PC3*10, y = PC4*10, label = label)) + 
  coord_cartesian(xlim = c(-6, 10), ylim = c(-8, 10)) +
  scale_color_manual(values = c("#ffa62b", "#2FC6B7")) +
  labs(title = "PCA per 10kb windows", 
       subtitle = "PCs 3 and 4",
       x = paste0("PC3 (", round(100*eigs[3]/sum(eigs), 1), "% explained var.)"),
       y = paste0("PC4 (", round(100*eigs[4]/sum(eigs), 1), "% explained var.)"))



```

```{r}
#SNP vs genes
p1 = ggplot(for_plot, aes(Short_variant_number, Coverage_frac_gene)) +
  geom_point(aes(col = CHR_type), alpha = .2) + 
  theme_bw() +
  geom_smooth(col = "grey40", method=lm)

summary(lm(Short_variant_number ~ Coverage_frac_gene, data = for_plot))


#TIPs vs genes
p2 = ggplot(for_plot, aes(TIP_count, Coverage_frac_gene)) +
  geom_point(aes(col = CHR_type), alpha = .2) + 
  theme_bw() +
  geom_smooth(col = "grey40", method=lm)

summary(lm(TIP_count ~ Coverage_frac_gene, data = for_plot))

cowplot::plot_grid(p1 + theme(legend.position = "none"), p2, rel_widths = c(1, 1.4))


```


Curious to see what happens if I remove windows in accessory chromosomes.
```{r filtered PCA2}
temp2 = data_PCA2 %>% 
  dplyr::select(-Mappability) %>%
  separate(Window, into = "CHR", sep = ":", remove = F, extra = "drop") %>%
  filter(as.numeric(CHR) < 14) %>%
  dplyr::select(-CHR)

temp = as.matrix(temp[,c(2:ncol(temp))])[, which(apply(as.matrix(temp[,c(2:ncol(temp))]), 2, var) != 0)]

per_win.pca = prcomp(temp, center = TRUE,scale. = TRUE)

#Making custom PCA plot
arrows = as_tibble(as.data.frame(per_win.pca$rotation)) %>%
  mutate(label = rownames(per_win.pca$rotation))

temp = as.tibble(cbind(temp2, as.data.frame(per_win.pca$x)))

for_plot = temp %>%
  separate(col = Window, into = c("CHR"), remove = F, extra = "drop", sep = ":") %>%
  mutate(CHR_type = ifelse(as.numeric(CHR) < 14, "Core", "Accessory"))
eigs <- per_win.pca$sdev^2

ggplot() + 
    geom_point(data = for_plot, aes(x = PC1, y = PC2, col = CHR_type), alpha = .6, shape = 1) + 
    theme_bw() + 
    geom_segment(data = arrows, aes(x = 0, y = 0, xend = PC1*10, yend = PC2*10 ),
                 arrow = arrow(length = unit(0.03, "npc"))) +
    geom_text(data = arrows, aes(x = PC1*10, y = PC2*10, label = label)) + 
  coord_cartesian(xlim = c(-6, 10), ylim = c(-8, 7)) +
  scale_color_manual(values = c("#ffa62b", "#2FC6B7")) +
  labs(title = "PCA per 10kb windows", 
       subtitle = "Two first PCs",
       x = paste0("PC1 (", round(100*eigs[1]/sum(eigs), 1), "% explained var.)"),
       y = paste0("PC2 (", round(100*eigs[2]/sum(eigs), 1), "% explained var.)"))

ggsave(paste0(fig_dir, "Fig1_PCA_per_10kb.pdf"), width = 20, height = 12, units = "cm")


ggplot() + 
    geom_point(data = for_plot, aes(x = PC3, y = PC4, col = CHR_type), alpha = .6, shape = 1) + 
    theme_bw() + 
    geom_segment(data = arrows, aes(x = 0, y = 0, xend = PC3*10, yend = PC4*10 ),
                 arrow = arrow(length = unit(0.03, "npc"))) +
    geom_text(data = arrows, aes(x = PC3*10, y = PC4*10, label = label)) + 
  coord_cartesian(xlim = c(-6, 10), ylim = c(-8, 10)) +
  scale_color_manual(values = c("#ffa62b", "#2FC6B7")) +
  labs(title = "PCA per 10kb windows", 
       subtitle = "PCs 3 and 4",
       x = paste0("PC3 (", round(100*eigs[3]/sum(eigs), 1), "% explained var.)"),
       y = paste0("PC4 (", round(100*eigs[4]/sum(eigs), 1), "% explained var.)"))



```
